---
title: "Import Kallisto Data / Pre-processing"
output: html_notebook
---

```{r, include=FALSE, message=FALSE, warning=FALSE}
library(AnnotationHub)
library(dplyr)
library(magrittr)
library(edgeR)
library(limma)
library(sleuth)
library(here)
library(readr)
library(tibble)
library(grid)
library(export)
library(reshape2)

# ggplot2 theme
theme_set(theme_bw())
```

## 1. Import Kallisto Counts

- **Kallisto** (v0.45.0) was run on Phoenix as follows: 
```{r engine='bash',eval=FALSE}
# Use local version of kallisto v0.45 as it's not installed on phoenix
kallisto=/data/biohub/local/kallisto_linux-v0.45.0/kallisto

# Kallisto index:
$kallisto index -i $z11/index/kallisto_GRCz11.cdna.all.idx $z11_cdna
KALISTOIDX=/data/biohub/Refs/zebrafish/GRCz11_Ensembl/kallisto_primaryOnly_GRCz11_index

# Run Kallisto for each trimmed fastq file
cd $TRIMDATA/fastq
TRIM=`ls ${SLURM_ARRAY_TASK_ID}_*.fastq.gz`

# Because the data is single-end, we need to specify length -l and s.d -s of the fragment length
# We use the defaults for kallisto which are 200nt and 20nt s.d. 
# mkdir -pv $QUANTDATA/transcriptLevel/${TRIM%%.fastq.gz}
OUTPUTDIR=/data/biohub/20170906_Morgan_Hypoxia/3_quantifiedData/kallisto_transcripts/${TRIM%%.fastq.gz}
$kallisto quant --genomebam --gtf $z11_gtf -t 16 --rf-stranded --single -b 100 -l 200 -s 20 -i $KALLISTOIDX -o $OUTPUTDIR $TRIM
```

### 1.1. catchKallisto()

- The **Kallisto** directories were imported into R using the `catchKallisto` function from *edgeR*:
```{r, eval=FALSE}
# Get paths to the Results directories generated by Kallisto.
kallistoResults <- "/Volumes/biohub/20170906_Morgan_Hypoxia/3_quantifiedData/transcriptLevel" %>% 
  file.path %>%
  list.dirs
kallistoResults <- kallistoResults[-1] #The first dir is the root dir, which can be removed. 
kallistoResults

############################################################################################################
# Import the kallisto results from the list of directories. 
# Takes a while to run so I've saved the rds

importedKallisto <- catchKallisto(kallistoResults)
importedKallisto %>% saveRDS(here("R","DE","data","importedKallisto.rds"))

#############################################################################################################
# Make colnames neater / not include the full path
importedKallisto$counts %<>% set_colnames(basename(colnames(.))) %>% 
  set_colnames(gsub(x=colnames(.), 
                    pattern="^.*_(6|24)(Q|P)_P(N|H)([1-4]).*$", 
                    replacement = "\\1\\2_P\\3\\4"))

str(importedKallisto)
```

### 1.2. Sample Information

- Load the sample spreadsheet including information like **Age**, **Genotype**, **Hypoxia/Normoxia**, and **Gender**. 
```{r }
samples <- read_csv(here("R","DE","data","samples.csv")) %>%
  mutate(
    Age = as.factor(as.character(Age)) %>% relevel(ref = "6"),
    Genotype = as.factor(Genotype) %>% relevel(ref = "wt"),
    Hypoxia = as.factor(as.character(as.numeric(Hypoxia))) %>% relevel(ref = "0"),
    Gender = as.factor(Gender) %>% relevel(ref = "female")
  ) %>% as.data.frame %>%
  set_rownames(.$NewName)%>%
  dplyr::mutate(id = gsub(x=id, pattern = "-", replacement = "_"))

# Add sample info to the filenames of the directories stored on Phoenix. 
sample_to_cov <- data.frame(path = kallistoResults) %>%
  mutate(basename = basename(as.character(path))) %>%
  mutate(id = gsub(x=basename,  pattern="^.*_(6|24)(Q|P)_P(N|H)([1-4]).*$", replacement = "\\1\\2_P\\3\\4")) %>% 
  left_join(samples, by = "id") %>%
  arrange(Age, Genotype, Hypoxia, Group) %>%
  dplyr::rename(sample = NewName) %>%
  dplyr::mutate(path = as.character(path))

sample_to_cov %>% as_tibble()
```

### 1.3. Transcript-level counts

- Create matrix of the transcript-level count estimates with new names from the samples table:
```{r}
counts_t <- importedKallisto$counts[, sample_to_cov$id] %>% set_colnames(sample_to_cov$sample)

counts_t %>% as.data.frame %>% rownames_to_column("transcript_id") %>% as_tibble
```

### 1.4. Gene-Transcript Annotation

- To summarise to gene-level counts, we first need to prepare a `data.frame` that maps from Ensembl transcript to gene IDs. 

```{r}
# Search for and load relevant AnnotationHub object:
ah <- AnnotationHub()
ah %>%
  subset(grepl("rerio", species)) %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]

# Get gene and transcript information:
genes <- genes(ensDb)
transcripts <- transcripts(ensDb)

mapTx2Gene <- transcripts %>% as.data.frame %>% dplyr::select(tx_id_version, gene_id)
mapTx2Gene %>% rownames_to_column("transcript_id")%>%as_tibble()
```

### 1.5. Gene-level counts

- Now we can summarise the transcript-level count estimates to gene-level count estimates by summing together the transcript-level counts corresponding to the same gene. We will also use the new names from the `samples` table. 

```{r}
counts_g <- counts_t %>%
  melt %>% 
  set_colnames(c("tx_id_version", "sample", "count")) %>% 
  left_join(mapTx2Gene) %>% 
  dcast(gene_id ~ sample, value.var = "count", fun.aggregate=sum)%>%
  magrittr::extract(,c("gene_id",sample_to_cov$sample)) 

counts_g %>% as_tibble
```

### 1.5. Export counts

- Finally, export out the relevant objects.

```{r, eval=FALSE}
#counts_t %>% saveRDS(here("R","DE","data","counts_t.rds"))
#counts_g %>% saveRDS(here("R","DE","data", "counts_g.rds"))
```


## 2. Prepare DGEList

- We will store gene and transcript count estimates as DGEList objects for differential gene/transcript analysis later. DGEList objects also contain gene/transcript annotation and sample annotation. 

- The DGEList for gene-level counts was created as follows: 

```{r eval=FALSE}
counts_for_dge <- counts_g %>% column_to_rownames("gene_id") %>% as.matrix

genes_for_dge <- data.frame(ensembl_gene_id = rownames(counts_for_dge)) %>%
  left_join(as.data.frame(genes) %>% dplyr::select(gene_id, gene_name, gene_biotype, entrezid, description), 
            by = c("ensembl_gene_id"="gene_id"))

samples_for_dge <- sample_to_cov %>% set_rownames(.$sample)

dge <- DGEList(counts = counts_for_dge, 
               genes = genes_for_dge, 
               samples = samples_for_dge, 
               remove.zeros = TRUE) %>%  
  calcNormFactors("TMM")

dge

#saveRDS(dge, here("R","DE","data","dge_g.rds"))
dge <- readRDS(here("R","DE","data","dge_g.rds"))
```

- The DGEList for transcript-level counts was created as follows:

```{r eval=FALSE}
counts_for_dge_t <- counts_t %>% as.data.frame %>% rownames_to_column("ensembl_transcript_id") %>%
  magrittr::extract(, c("ensembl_transcript_id", sample_to_cov$sample)) %>%
  column_to_rownames("ensembl_transcript_id") %>% as.matrix

transcripts_for_dge_t <- data.frame(ensembl_transcript_id = rownames(counts_t)) %>%
  left_join(mapTx2Gene, by = c("ensembl_transcript_id"="tx_id_version")) %>%
  left_join(as.data.frame(genes) %>% dplyr::select(gene_id, gene_name, gene_biotype, entrezid, description), 
            by = c("gene_id"))%>%
  dplyr::rename(ensembl_gene_id = gene_id)

dge_t <- DGEList(counts = counts_for_dge_t,
                 genes = transcripts_for_dge_t,
                 samples = samples_for_dge, remove.zeros = TRUE) %>%
  calcNormFactors("TMM")

dge_t

# saveRDS(dge_t, here("R","DE","data","dge_t.rds"))
dge_t <- readRDS(here("R","DE","data","dge_t.rds"))
```


## 3. Visualisation

We will use Principal Component Analysis to get an idea of the similarity of samples / libraries based on their gene and transcript expression. 

### 3.1. Gene level PCA

```{r}
pca_analysis1 <- prcomp(t(cpm(dge, log=TRUE)))

pca_geneExpression <- pca_analysis1$x %>% magrittr::extract(, c("PC1", "PC2")) %>%
  as.data.frame %>%
  rownames_to_column("samples") %>%
  left_join((dge$samples %>% rownames_to_column("samples")), by="samples") %>%
  ggplot(aes(x=PC1, y = PC2, colour = Hypoxia, shape = paste0(Genotype, "_", Age))) +
  geom_point(alpha = 0.7,size=3) + 
  scale_shape_manual(values = c(0,1,15,16),
                     labels = c("Q96K97/+, 24 months", "Q96K97/+, 6 months", "+/+, 24 months", "+/+, 6 months")) +
  theme(aspect.ratio = 1) +
  scale_colour_manual(values = c("red","cornflowerblue"), labels = c("Hypoxia","Normoxia")) +
    labs(x = "Principal Component 1 (19.15%)", y = "Principal Component 2 (6.0%)", 
         colour = "Oxygen Level", shape = "Genotype and Age")

summary(pca_analysis1)
screeplot(pca_analysis1)
pca_geneExpression

# export::graph2ppt(pca_geneExpression, file = here("R","DE","fig","pca_geneExpression"))
```

### 3.2. Transcript level PCA:

```{r}
pca_analysis2 <- prcomp(t(cpm(dge_t, log=TRUE)))

summary(pca_analysis2)

pca_geneExpression2 <- pca_analysis2$x %>% magrittr::extract(, c("PC1", "PC2")) %>%
  as.data.frame %>%
  rownames_to_column("samples") %>%
  left_join((dge$samples %>% rownames_to_column("samples")), by="samples") %>%
  ggplot(aes(x=PC1, y = PC2, colour = Hypoxia, shape = paste0(Genotype, "_", Age))) +
  geom_point(alpha = 0.7,size=3) + 
  scale_shape_manual(values = c(0,1,15,16),
                     labels = c("Q96K97/+, 24 months", "Q96K97/+, 6 months", "+/+, 24 months", "+/+, 6 months")) +
  theme(aspect.ratio = 1) +
  scale_colour_manual(values = c("red","cornflowerblue"), labels = c("Hypoxia","Normoxia")) +
    labs(x = "Principal Component 1 (9.4%)", y = "Principal Component 2 (7.8%)", 
         colour = "Oxygen Level", shape = "Genotype and Age")

summary(pca_analysis2)
screeplot(pca_analysis2)
pca_geneExpression2


```

## 4. Filtering

### 4.1. Gene-level filtering

We chose the threshold of at least 1 cpm in at least 4 samples based on the following density plots. 

```{r}
keepTheseGenes <- (rowSums(cpm(dge) > 1) >= 4) 

A <- dge %>% 
  cpm(log = TRUE) %>% 
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("A. Before filtering") +
  labs(x = "logCPM", y = "Density")

B <- dge %>% 
  cpm(log = TRUE) %>% 
  magrittr::extract(keepTheseGenes,) %>%
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("B. After filtering")+
  labs(x = "logCPM", y = "Density")

grid.newpage()
vp1 <- viewport(x = 0, y = 0, width = 0.5, height = 1, just = c(0, 0))
vp2 <- viewport(x = 0.5, y = 0, width = 0.5, height = 1, just = c(0,0))
print(A, vp = vp1)
print(B, vp  = vp2)
```

- Using this threshold, we filtered out `r table(keepTheseGenes)[[1]]` genes from the original `r length(keepTheseGenes)` genes, giving the remaining `r table(keepTheseGenes)[[2]]` to be used in the analysis. 
```{r}
dge <- dge[keepTheseGenes,,keep.lib.sizes = FALSE] 
# saveRDS(dge, here("R","DE","data","dge_g_filtered.rds"))
```

### 4.2. Transcript level filtering

The threshold for filtering (at least 1 cpm in at least 4 samples) was chosen to be consistent with the gene level filtering.

```{r}
keepTheseTx <- (rowSums(cpm(dge_t) > 1) >= 4) 

A <- dge_t %>% 
  cpm(log = TRUE) %>% 
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("A. Before filtering") +
  labs(x = "logCPM", y = "Density")

B <- dge_t %>% 
  cpm(log = TRUE) %>% 
  magrittr::extract(keepTheseTx,) %>%
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("B. After filtering")+
  labs(x = "logCPM", y = "Density")

grid.newpage()
vp1 <- viewport(x = 0, y = 0, width = 0.5, height = 1, just = c(0, 0))
vp2 <- viewport(x = 0.5, y = 0, width = 0.5, height = 1, just = c(0,0))
print(A, vp = vp1)
print(B, vp  = vp2)
```

- Using this threshold, we filtered out `r table(keepTheseTx)[[1]]` transcripts from the original `r length(keepTheseTx)` transcripts, giving the remaining `r table(keepTheseTx)[[2]]` to be used in the analysis. 
```{r}
dge_t <- dge_t[keepTheseTx,,keep.lib.sizes = FALSE] 
saveRDS(dge_t, here("R","DE","data","dge_t_filtered.rds"))
```

## 5. Session Info

```{r}
sessionInfo()
```

