---
title: "GSEA overlap with IRE genes - Mouse "
output: html_notebook
---

```{r include=FALSE}
library(GSEABase)
library(dplyr)
library(readr)
library(magrittr)
library(tibble)
library(reshape2)
library(fgsea)
library(ggplot2)
library(org.Hs.eg.db)
library(org.Dr.eg.db)
library(org.Mm.eg.db)
library(biomaRt)
library(limma)
library(stringr)
library(openxlsx)
library(pheatmap)
library(here)
```


## 0. Background

- In the `GSEA_Overlap_with_IREGenes.Rmd` notebook, Fisher's exact test was used to determine whether MSigDB gene sets were enriched in the zebrafish predicted IRE gene sets we defined. 
- Now we are going to repeat the same analysis, but for the mouse predicted IRE gene sets.

## 1. Import gene set collections from MSigDB

- As in the original analysis, the following collections from MSigDB will be imported:
  - Hallmark gene sets
  - C2 - Curated gene sets
  - C3 - Motif gene sets
  - C4 - Gene ontology gene sets
- As human downloads are readily available from the MSigDB, we will not need to convert between species. 
- Mouse is not available from MSigDB. [WEHI](http://bioinf.wehi.edu.au/software/MSigDB/) have a download but it's for version 5.2 so to maintain consistency with the human one, we will do a fresh conversion. 
- However, we will need to convert the ID from entrezgene into Ensembl ID to remain consistent across all of our analyses. 

```{r}
genesetsDir <- here::here("R", "GSEA", "data", "human")

h <- getGmt(file.path(genesetsDir, "h.all.v7.0.entrez.gmt"))    # Hallmark gene sets
c1 <- getGmt(file.path(genesetsDir, "c1.all.v7.0.entrez.gmt"))  # positional gene sets
c2 <- getGmt(file.path(genesetsDir, "c2.all.v7.0.entrez.gmt"))  # curated gene sets
c3 <- getGmt(file.path(genesetsDir, "c3.all.v7.0.entrez.gmt"))  # motif gene sets
c4 <- getGmt(file.path(genesetsDir, "c4.all.v7.0.entrez.gmt"))  # computational gene sets
c5 <- getGmt(file.path(genesetsDir, "c5.all.v7.0.entrez.gmt"))  # GO gene sets
c6 <- getGmt(file.path(genesetsDir, "c6.all.v7.0.entrez.gmt"))  # oncogenic signatures gene sets
c7 <- getGmt(file.path(genesetsDir, "c7.all.v7.0.entrez.gmt"))  # immunologic signatures gene sets


# Convert each gene sets to list where the name of each list is the gene set
# name and the list items are the entrezgenes. 
h_l <- geneIds(h) %>% as.list
c1_l <- geneIds(c1) %>% as.list
c2_l <- geneIds(c2) %>% as.list
c3_l <- geneIds(c3) %>% as.list
c4_l <- geneIds(c4) %>% as.list
c5_l <- geneIds(c5) %>% as.list
c6_l <- geneIds(c6) %>% as.list
c7_l <- geneIds(c7) %>% as.list

# Bind the list of gene sets so that each list becomes a data.frame.
h_df <- lapply(h_l, cbind)
c1_df <- lapply(c1_l, cbind)
c2_df <- lapply(c2_l, cbind)
c3_df <- lapply(c3_l, cbind)
c4_df <- lapply(c4_l, cbind)
c5_df <- lapply(c5_l, cbind)
c6_df <- lapply(c6_l, cbind)
c7_df <- lapply(c7_l, cbind)
```

## 1.1. Convert human entrezgene -> human ensembl

```{r}
human_entrez2Ens <- as.data.frame(org.Hs.egENSEMBL) %>%
  set_colnames(c("human_entrezgene", "human_ensembl"))

map_human_entrez2Ens <- function(x) {
  x %>% 
    as.data.frame %>% 
    set_colnames("human_entrezgene") %>%
    left_join(human_entrez2Ens, by = "human_entrezgene") %>% 
    dplyr::filter(complete.cases(.)) %>%
    dplyr::select(-human_entrezgene) %>%
    as.list %>%
    unname %>%
    .[[1]] %>%
    unique
}

h_mapped <- lapply(h_df, map_human_entrez2Ens)
c1_mapped <- lapply(c1_df, map_human_entrez2Ens)
c2_mapped <- lapply(c2_df, map_human_entrez2Ens) 
c3_mapped <- lapply(c3_df, map_human_entrez2Ens)
c4_mapped <- lapply(c4_df, map_human_entrez2Ens)
c5_mapped <- lapply(c5_df, map_human_entrez2Ens)
c6_mapped <- lapply(c6_df, map_human_entrez2Ens)
c7_mapped <- lapply(c7_df, map_human_entrez2Ens)

h_mapped %>% saveRDS(file.path(genesetsDir, "ens_h_mapped.rds"))
c1_mapped %>% saveRDS(file.path(genesetsDir, "ens_c1_mapped.rds"))
c2_mapped %>% saveRDS(file.path(genesetsDir, "ens_c2_mapped.rds"))
c3_mapped %>% saveRDS(file.path(genesetsDir, "ens_c3_mapped.rds"))
c4_mapped %>% saveRDS(file.path(genesetsDir, "ens_c4_mapped.rds"))
c5_mapped %>% saveRDS(file.path(genesetsDir, "ens_c5_mapped.rds"))
c6_mapped %>% saveRDS(file.path(genesetsDir, "ens_c6_mapped.rds"))
c7_mapped %>% saveRDS(file.path(genesetsDir, "ens_c7_mapped.rds"))

h_mapped <- readRDS(file.path(genesetsDir, "ens_h_mapped.rds"))
c1_mapped <- readRDS(file.path(genesetsDir, "ens_c1_mapped.rds"))
c2_mapped <- readRDS(file.path(genesetsDir, "ens_c2_mapped.rds"))
c3_mapped <- readRDS(file.path(genesetsDir, "ens_c3_mapped.rds"))
c4_mapped <- readRDS(file.path(genesetsDir, "ens_c4_mapped.rds"))
c5_mapped <- readRDS(file.path(genesetsDir, "ens_c5_mapped.rds"))
c6_mapped <- readRDS(file.path(genesetsDir, "ens_c6_mapped.rds"))
c7_mapped <- readRDS(file.path(genesetsDir, "ens_c7_mapped.rds"))

```

## 1.2. Convert human entrezgene -> mouse ensembl

```{r}
# Mouse entrez and ensembl ids:
mEns <- org.Mm.egENSEMBL %>%
  as.data.frame %>%
  set_colnames(c("m_entrezgene","m_ensembl"))

# Create a data.frame to map between human entrezgenes and zebrafish ensembl IDs.
# BioMart only includes homolog mappings for ensembl IDs which is why we need to 
# retrieve human ensembl IDs, then join to the humanEntrezEns data.frame,
# in order to get the desired human entrezgenes to zebrafish ensembl ID mapping. 
mMart <- useMart("ENSEMBL_MART_ENSEMBL", "mmusculus_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "hsapiens_homolog_ensembl_gene")
mAndHumanEnsGenes <- getBM(getFromBiomart, values = unique(mEns$m_ensembl), mart = mMart) %>%
  set_colnames(c("zebrafish_ensembl", "human_ensembl")) %>%
  left_join(human_entrez2Ens, by = "human_ensembl") %>%
  dplyr::select(-human_ensembl) %>% 
  dplyr::filter(complete.cases(.))

# mToHu <- getBM(c("ensembl_gene_id", "hsapiens_homolog_ensembl_gene"),
#                     values = unique(mEns$m_ensembl), mart = mMart) %>%
#   set_colnames(c("m_ensembl","hu_ensembl"))

```

```{r}
mapHumanGS2Mouse<- function(x) {
  x %>% 
    as.data.frame %>% 
    set_colnames("human_entrezgene") %>%
    left_join(mAndHumanEnsGenes, by = "human_entrezgene") %>% 
    dplyr::filter(complete.cases(.)) %>%
    dplyr::select(-human_entrezgene) %>%
    as.list %>%
    unname %>%
    .[[1]] %>%
    unique
}

h_mapped_m <- lapply(h_df, mapHumanGS2Mouse)
c1_mapped_m <- lapply(c1_df, mapHumanGS2Mouse)
c2_mapped_m <- lapply(c2_df, mapHumanGS2Mouse) 
c3_mapped_m <- lapply(c3_df, mapHumanGS2Mouse)
c4_mapped_m <- lapply(c4_df, mapHumanGS2Mouse)
c5_mapped_m <- lapply(c5_df, mapHumanGS2Mouse)
c6_mapped_m <- lapply(c6_df, mapHumanGS2Mouse)
c7_mapped_m <- lapply(c7_df, mapHumanGS2Mouse)

h_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_h_mapped_m.rds"))
c1_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c1_mapped_m.rds"))
c2_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c2_mapped_m.rds"))
c3_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c3_mapped_m.rds"))
c4_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c4_mapped_m.rds"))
c5_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c5_mapped_m.rds"))
c6_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c6_mapped_m.rds"))
c7_mapped_m %>% saveRDS(file.path(genesetsDir, "ens_c7_mapped_m.rds"))

h_mapped_m <- readRDS(file.path(genesetsDir, "ens_h_mapped_m.rds"))
c1_mapped_m <- readRDS(file.path(genesetsDir, "ens_c1_mapped_m.rds"))
c2_mapped_m <- readRDS(file.path(genesetsDir, "ens_c2_mapped_m.rds"))
c3_mapped_m <- readRDS(file.path(genesetsDir, "ens_c3_mapped_m.rds"))
c4_mapped_m <- readRDS(file.path(genesetsDir, "ens_c4_mapped_m.rds"))
c5_mapped_m <- readRDS(file.path(genesetsDir, "ens_c5_mapped_m.rds"))
c6_mapped_m <- readRDS(file.path(genesetsDir, "ens_c6_mapped_m.rds"))
c7_mapped_m <- readRDS(file.path(genesetsDir, "ens_c7_mapped_m.rds"))
```



## 2. Import IRE gene set

```{r}
mouseIreGenes_h<-readRDS(here::here("R/IREGenes/data/mouseIreGenes_h.rds"))
humanIreGenes<-readRDS(here::here("R/IREGenes/data/human_ireGenes.rds"))
zebrafishIreGenes_h<-readRDS(here::here("R/IREGenes/data/zebrafishIreGenes_h.rds"))
```

## 3. Human: Overlap with GSEA Analysis 

```{r}
h_tib <- h_mapped %>% tibble %>% set_colnames(c("ids")) %>% mutate(geneset = names(h_mapped), source = "h") 
c2_tib <- c2_mapped %>% tibble %>% set_colnames(c("ids")) %>% mutate(geneset = names(c2_mapped), source = "c2")
c3_tib <- c3_mapped %>% tibble %>% set_colnames(c("ids")) %>% mutate(geneset = names(c3_mapped), source = "c3")
c5_tib <- c5_mapped %>% tibble %>% set_colnames(c("ids")) %>% mutate(geneset = names(c5_mapped), source = "c5")

gs <- bind_rows(h_tib, c2_tib, c3_tib, c5_tib)

gs
```

```{r}
humanIreGenes%>%str

allIREGenes <- c(humanIreGenes$ire3_all, humanIreGenes$ire5_all)

allIREGenes %>% head
```

### 3.1. Compute gene overlap with IRE genesets

We wish to see the overlap between the genesets in `gs` with the `humanIreGenes` (including 3' and 5' IRE genes), 
as well as separately, with the 3' IRE genes and 5' IRE genes. We will add this information into the `gs` tibble.


```{r}
gs %<>% rowwise() %>% mutate(
  n = length(ids),  # Number of genes in the geneset
  
  n_with_ire = sum(ids %in% allIREGenes),  # Number of genes in the gene set which have 3' or 5' predicted IREs 
  n_without_ire = (n - n_with_ire),
  universe_with_ire = sum(human_entrez2Ens$human_ensembl %in% allIREGenes & !(human_entrez2Ens$human_ensembl %in% ids)), 
  universe_without_ire = (length(human_entrez2Ens$human_ensembl) - universe_with_ire),
  
  n_with_ire3 = sum(ids %in% humanIreGenes$ire3_all),  # Number of genes in the gene set which have 3' predicted IREs
  n_without_ire3= (n - n_with_ire),
  universe_with_ire3 = sum(human_entrez2Ens$human_ensembl %in% humanIreGenes$ire3_all & !(human_entrez2Ens$human_ensembl %in% ids)), 
  universe_without_ire3 = (length(human_entrez2Ens$human_ensembl) - universe_with_ire3),
  
  n_with_ire5 = sum(ids %in% humanIreGenes$ire5_all),  # Number of genes in the gene set which have 5' predicted IREs
  n_without_ire5= (n - n_with_ire),
  universe_with_ire5 = sum(human_entrez2Ens$human_ensembl %in% humanIreGenes$ire5_all & !(human_entrez2Ens$human_ensembl %in% ids)), 
  universe_without_ire5 = (length(human_entrez2Ens$human_ensembl) - universe_with_ire5)
) %>% ungroup()

gs
```

```{r, eval=FALSE}
gs_mat <- gs %>% 
  dplyr::select(n_with_ire,
                n_without_ire, 
                universe_with_ire, 
                universe_without_ire) %>% 
  apply(X = ., MARGIN = 1, FUN = function(x){
    x %>% 
      matrix(2,2) %>% 
      t %>% 
      list()
  }) %>% set_names(gs$geneset) %>% 
  lapply(function(x){
    x %>% .[[1]]
  })

gs_mat3 <- gs %>% 
  dplyr::select(n_with_ire3,
                n_without_ire3, 
                universe_with_ire3, 
                universe_without_ire3) %>% 
  apply(X = ., MARGIN = 1, FUN = function(x){
    x %>% 
      matrix(2,2) %>% 
      t %>% 
      list()
  }) %>% set_names(gs$geneset) %>% 
  lapply(function(x){
    x %>% .[[1]]
  })

gs_mat5 <- gs %>% 
  dplyr::select(n_with_ire5,
                n_without_ire5, 
                universe_with_ire5, 
                universe_without_ire5) %>% 
  apply(X = ., MARGIN = 1, FUN = function(x){
    x %>% 
      matrix(2,2) %>% 
      t %>% 
      list()
  }) %>% set_names(gs$geneset) %>% 
  lapply(function(x){
    x %>% .[[1]]
  })

```
```{r}
gs_mat[1:3]
gs_mat3[1:3]
gs_mat5[1:3]
```
```{r}
fisher_res <- gs_mat %>% lapply(function(x){
  x %>% fisher.test()
})

fisher_res3 <- gs_mat3 %>% lapply(function(x){
  x %>% fisher.test()
})

fisher_res5 <- gs_mat5 %>% lapply(function(x){
  x %>% fisher.test()
})

fisher_res_p <- fisher_res %>% lapply(function(x){x$p.value})
fisher_res_p3 <- fisher_res3 %>% lapply(function(x){x$p.value})
fisher_res_p5 <- fisher_res5 %>% lapply(function(x){x$p.value})

gs %<>% 
  mutate(fisher_p = fisher_res_p%>%unlist%>%unname,
         fisher_p_3 = fisher_res_p3%>%unlist%>%unname,
         fisher_p_5 = fisher_res_p5%>%unlist%>%unname) %>%
  mutate(fdr = p.adjust(fisher_p, "fdr"),
         fdr_3 = p.adjust(fisher_p_3, "fdr"),
         fdr_5 = p.adjust(fisher_p_5, "fdr"))
```

### Calculate Expected and Observed

For each gene set, we will calculate the number of genes **expected** to have IREs (based on the background proportion) and **observed** value. 
This information will be added to the `gs` object.

```{r eval=FALSE}
gs %<>% rowwise() %>% mutate(
  exp_allIRE = (universe_with_ire / universe_without_ire)*n_without_ire,
  obs_allIRE = n_with_ire,
  obs_greater_than_exp_allIRE = obs_allIRE > exp_allIRE,
  
  exp_ire3 = (universe_with_ire3 / universe_without_ire3)*n_without_ire3,
  obs_ire3 = n_with_ire3,
  obs_greater_than_exp_ire3 = obs_ire3 > exp_ire3,
  
  exp_ire5 = (universe_with_ire5 / universe_without_ire5)*n_without_ire5,
  obs_ire5 = n_with_ire5,
  obs_greater_than_exp_ire5 = obs_ire5 > exp_ire5
) 
```

```{r}
gs %>%
  dplyr::select(geneset, contains("exp"), starts_with("obs"), ) %>% ungroup()
```

## 6. Results

The following table shows the top 50 gene sets enriched in IRE genesets, ranked by Fisher's exact test *p*-values:

```{r}
gs %>% ungroup() %>%
  arrange(fisher_p) %>%
  dplyr::select(geneset, contains("fdr"), n_with_ire, n, starts_with("obs_greater")) 

#gs %>% saveRDS(here::here("R/GSEA/data/ora/human_gs.rds"))
gs <- readRDS(here::here("R/GSEA/data/ora/human_gs.rds"))
```

Sorted by genesets most enriched in 3' IRE genes:
```{r}
gs %>% ungroup %>% arrange(fisher_p_3) %>% dplyr::select(geneset, contains("fdr"), n_with_ire3, n, starts_with("obs_greater")) %>% head(50)
```

Sorted by genesets most enriched in 5' IRE genes:
```{r}
gs %>% ungroup %>% arrange(fisher_p_5) %>% dplyr::select(geneset, contains("fdr"), n_with_ire5, n, starts_with("obs_greater")) %>% head(50)
```

## 7. Visualisation (Stacked bar chart)

The following stacked bar chart shows the overlap between the top ~20 genesets and the predicted-IRE genesets.

```{r }
overlapDf <- gs %>% arrange(fisher_p) %>% dplyr::filter(fdr < 0.01 | fdr_3 < 0.01 | fdr_5 < 0.01) %>%
  slice(1:15) %>%
  dplyr::filter(obs_greater_than_exp_allIRE == TRUE | obs_greater_than_exp_ire3 == TRUE | obs_greater_than_exp_ire5 == TRUE) %>%
  dplyr::select(geneset, source, n_with_ire3, n_with_ire5, n_without_ire) %>%
  bind_rows(data.frame(
    geneset = c("- All Predicted IREs"),
    source = c("sires"),
    n_with_ire3 = c(length(humanIreGenes$ire3_all)),
#                    ireUtr3 %>% as.data.frame %>% dplyr::filter(quality == "High") %>% use_series("gene_id") %>% unique %>% length),
    n_with_ire5 = c(length(humanIreGenes$ire5_all)),
#                    ireUtr5 %>% as.data.frame %>% dplyr::filter(quality == "High") %>% use_series("gene_id") %>% unique %>% length),
    n_without_ire = c(0)
  )) %>% dplyr::mutate(geneset = paste0(geneset, " ", source)) %>%
  dplyr::mutate(geneset = gsub(x=geneset, pattern = "_", replacement = " ")) %>%
  dplyr::select(-source) %>% melt

totals <- overlapDf %>% 
  dplyr::arrange(geneset) %>% 
  group_by(geneset) %>% 
  summarise(sum = sum(value)) %>% 
  arrange(desc(sum))


overlapPlot <- overlapDf %>% 
  left_join(totals) %>%
  arrange(desc(sum)) %>%
  mutate(geneset = factor(geneset, levels = unique(as.character(geneset)))) %>%
  ggplot(aes(x = str_to_title(stringr::str_wrap(geneset, 28)) %>% factor(., levels = unique(as.character(.))), 
                     y = value, 
                     fill = variable)) + 
  geom_col(width = 0.4) +
  theme(aspect.ratio = 1.2, 
        axis.text.x =  element_text(size= 35), 
        axis.title.x = element_text(size= 30),
        axis.title.y = element_text(size = 30), 
        legend.text = element_text(size = 20), 
        legend.title = element_text(size=30),
        axis.text.y = element_text(color = "grey20", size = 13)) + 
  coord_flip() +  # Axis labels too long to be readable so this helps.
  scale_fill_manual(values=c("#e11757", "#3767b3", "#dddddd"), labels =c("3' IRE", "5' IRE", "No IRE")) +
  labs(y = "Number of genes", x = "Gene set", fill = "Gene\nhas\nIRE?") +
  theme_bw()

overlapPlot

export::graph2ppt(overlapPlot, here("R","GSEA","fig","overlapPlot_human_top15"))
```

## 8. Visualisation (network overlap)

- The stacked bar chart doesn't indicate to us whether the IRE genes in 
each gene set are actually shared or if they're unique. Overlapping sets 
through Venn diagrams or UpSet plots would be appropriate for showing this, but 
I'm going to use a network plot for the following reasons:

    - There are ~15 sets which are to be plotted. Although UpSet copes OK with 
    this, the main point which I want to emphasise is how many IRE genes 
    are UNIQUE to the predicted IRE gene sets? 
    - Network layout algorithms can be informative in showing us exactly this 
    along with giving an appreciation of the scale (e.g. how many genes 
    are involved relatively in each gene set / are shared). 

- To use Gephi to produce a network plot, we need a **nodes** table and 
an **edges** table. Here I will produce the nodes table using the top 
15 gene sets ranked by *p*-value. 

- We will also append the predicted IRE gene sets and the "Hallmark 
Heme Metabolism" gene set as this one is considered like a "gold standard" and 
the first point for testing for many as it's included in the Hallmark 
collection. 

```{r}
# Include the following top ranked gene sets (~15) in the network 
nodes <- gs %>% arrange(fisher_p) %>% 
  dplyr::slice(1:15) %>%
  dplyr::filter(obs_greater_than_exp_allIRE == TRUE | 
                  obs_greater_than_exp_ire3 == TRUE | 
                  obs_greater_than_exp_ire5 == TRUE) %>%
  ungroup

# Append information about 3' and 5' predicted IRE genesets 
nodes %<>% bind_rows(
   tibble(
     ids = c(list(humanIreGenes$ire3_all, humanIreGenes$ire5_all)),
     geneset = c("Predicted 3' IRE genes", "Predicted 5' IRE genes"),
     source = c("sires","sires"),
     n = c(length(humanIreGenes$ire3_all), length(humanIreGenes$ire5_all))
   )
 ) 

# Append the Hallmark Heme Metabolism geneset
# Although this gene set wasnt significantly enriched in 
# IRE-containing genes, it is probably the most comprehensive
# gene set specifically on heme metabolism, and combines info 
# from various studies. 
nodes %<>% bind_rows(
  h_tib %>% filter(geneset == "HALLMARK_HEME_METABOLISM") %>% 
    mutate(n = length(ids[[1]]))
 ) 

nodes
```

- Below, we create a data.frame to hold information about gene intersections 
from each gene set. 

```{r}
gsComb <- combn(nodes$geneset, m = 2) %>% 
  t %>% 
  as.data.frame %>%
  set_colnames(c("geneset", "geneset_2b")) %>%
  left_join(nodes%>%dplyr::select(ids, geneset), by = "geneset") %>%
  dplyr::rename(geneset_nm = geneset,
                geneset = geneset_2b) %>%
  left_join(nodes%>%dplyr::select(ids, geneset), by = "geneset") %>%
  as_tibble %>%
  mutate(common_ids = map2(ids.x, ids.y, ~intersect(.x,.y)))

gsComb 
```

- We can extract genes which appear the most often in genesets as follows:

```{r}
gsComb$common_ids%>%
  unlist %>% 
  table %>% 
  as.data.frame%>%
  arrange(desc(Freq)) %>%
  set_colnames(c("ensembl_gene_id", "freq"))  %>% head
```

- The final nodes table for gephi will contain

    - All genes in the genesets
    - The geneset names

```{r}
pathways <- nodes$geneset %>% as.data.frame %>% set_colnames("label")
genes <- nodes$ids %>% unlist %>% unique %>% as.data.frame %>% set_colnames("label")
nodesDf <- full_join(pathways, genes, by = "label") %>% rowid_to_column("id")

# nodesDf %>%
#   mutate(text = ifelse(id < 18, label, NA)) %>%
#   mutate(size = ifelse(id < 16, 2, 1)) %>%
#   mutate(colour = ifelse(id < 16, rainbow(26)[id], NA)) 

head(nodesDf,20)

nodesDf2 <- nodesDf %>%
  mutate(
    ire = case_when(
      id < 16 ~ " ",
      id == 18 ~ " ",
      id == 16 ~ "3",
      id == 17 ~ "5",
      label %in% humanIreGenes$ire3_all ~ "3",
      label %in% humanIreGenes$ire5_all ~ "5",
      !(label %in% allIREGenes) ~ "no IRE"
    )
  )%>%
  dplyr::select(-id) %>% dplyr::rename(Id = label) 

head(nodesDf2, 30) 
 nodesDf2 %>% write_tsv(here("R","GSEA","data","human_network","nodes.tsv"))
```

- Edges will be between:

    - The genesets and all genes within the geneset 
    - Genes in multiple genesets

```{r}
edgeDf <- nodes$ids %>% set_names(nodes$geneset) %>% plyr::ldply(data.frame) %>% 
  set_colnames(c("pathway", "gene")) %>%
  left_join(nodesDf, by = c("pathway"="label")) %>%
  dplyr::rename(from = id) %>%
  left_join(nodesDf, by = c("gene"="label")) %>%
  dplyr::rename(to=id) %>%
  #dplyr::select(from, to)
  dplyr::select(pathway, gene) %>%
  set_colnames(c("Source","Target"))
edgeDf %>% head(20)

write_tsv(edgeDf, here::here("R/GSEA/data/human_network/edges.tsv"))

```

## 9. Session Info

```{r}
sessionInfo()
```


