---
title: "GSEA Analysis"
output: html_notebook
---

```{r Setup, include=FALSE,message=FALSE,warning=FALSE}
# Load packages:
library(GSEABase)
library(dplyr)
library(readr)
library(magrittr)
library(tibble)
library(reshape2)
library(fgsea)
library(ggplot2)
library(org.Hs.eg.db)
library(org.Dr.eg.db)
library(biomaRt)
library(limma)
library(here)
library(stringr)
library(openxlsx)
```


## 0. Introduction

The purpose of this analysis is to determine enriched gene sets in the Q96K97 dataset. 

In this analysis, we will need the voom object, design matrix, and contrasts matrix previously made in the differential gene expression analysis using *limma*.

```{r}
v <- readRDS(here("R","DE","data","voomData_g.rds"))  # Voom object
design <- readRDS(here("R","DE","data","design_g.rds"))  # Design matrix
contrasts <- readRDS(here("R","DE","data","contrasts_g.rds"))  # Contrasts matrix
```



## 1. Import gene sets 

The .gmt files from GSEA MSigDB were downloaded and imported into R. 

```{r}
genesetsDir <- here("R", "GSEA", "data")

h <- getGmt(file.path(genesetsDir, "h.all.v6.0.entrez.gmt"))    # Hallmark gene sets
c1 <- getGmt(file.path(genesetsDir, "c1.all.v6.0.entrez.gmt"))  # positional gene sets
c2 <- getGmt(file.path(genesetsDir, "c2.all.v6.0.entrez.gmt"))  # curated gene sets
c3 <- getGmt(file.path(genesetsDir, "c3.all.v6.0.entrez.gmt"))  # motif gene sets
c4 <- getGmt(file.path(genesetsDir, "c4.all.v6.0.entrez.gmt"))  # computational gene sets
c5 <- getGmt(file.path(genesetsDir, "c5.all.v6.0.entrez.gmt"))  # GO gene sets
c6 <- getGmt(file.path(genesetsDir, "c6.all.v6.0.entrez.gmt"))  # oncogenic signatures gene sets
c7 <- getGmt(file.path(genesetsDir, "c7.all.v6.0.entrez.gmt"))  # immunologic signatures gene sets

# Convert each gene sets to list where the name of each list is the gene set
# name and the list items are the entrezgenes. 
h_l <- geneIds(h) %>% as.list
c1_l <- geneIds(c1) %>% as.list
c2_l <- geneIds(c2) %>% as.list
c3_l <- geneIds(c3) %>% as.list
c4_l <- geneIds(c4) %>% as.list
c5_l <- geneIds(c5) %>% as.list
c6_l <- geneIds(c6) %>% as.list
c7_l <- geneIds(c7) %>% as.list

# Bind the list of gene sets so that each list becomes a data.frame.
h_df <- lapply(h_l, cbind)
c1_df <- lapply(c1_l, cbind)
c2_df <- lapply(c2_l, cbind)
c3_df <- lapply(c3_l, cbind)
c4_df <- lapply(c4_l, cbind)
c5_df <- lapply(c5_l, cbind)
c6_df <- lapply(c6_l, cbind)
c7_df <- lapply(c7_l, cbind)
```


## 2. Convert human entrezgenes in MSigDB gene sets to zebrafish ensembl IDs

The IDs in the MSigDB gene sets are human entrezgene IDs which need to be converted into zebrafish Ensembl gene IDs. We will prepare a data.frame `zebAndHumanEnsGenes` to map between the IDs.

```{r}
# Human entrez and ensembl IDs:
humanEntrezEns <- org.Hs.egENSEMBL %>% 
  as.data.frame %>%
  set_colnames(c("human_entrezgene", "human_ensembl"))

# Zebrafish ensembl IDs:
zebEns <- org.Dr.egENSEMBL %>%
  as.data.frame %>%
  set_colnames(c("zeb_entrezgene", "zeb_ensembl"))

# Create a data.frame to map between human entrezgenes and zebrafish ensembl IDs.
# BioMart only includes homolog mappings for ensembl IDs which is why we need to 
# retrieve human ensembl IDs, then join to the humanEntrezEns data.frame,
# in order to get the desired human entrezgenes to zebrafish ensembl ID mapping. 
zebMart <- useMart("ENSEMBL_MART_ENSEMBL", "drerio_gene_ensembl")
getFromBiomart <- c("ensembl_gene_id", "hsapiens_homolog_ensembl_gene")
zebAndHumanEnsGenes <- getBM(getFromBiomart, values = unique(zebEns$zeb_ensembl), mart = zebMart) %>%
  set_colnames(c("zebrafish_ensembl", "human_ensembl")) %>%
  left_join(humanEntrezEns, by = "human_ensembl") %>%
  dplyr::select(-human_ensembl) %>% 
  dplyr::filter(complete.cases(.))

```

This mapping in `zebAndHumanEnsGenes` contains `r (zebAndHumanEnsGenes %>% lapply(function(x){x%>%unique%>%length}))[[1]]` zebrafish Ensembl IDs to `r (zebAndHumanEnsGenes %>% lapply(function(x){x%>%unique%>%length}))[[2]]` human entrezgene IDs. 

The following function is applied to the gene sets to map the IDs in the gene sets from human entrezgene to zebrafish ensembl. 

```{r, eval=FALSE,message=FALSE,warning=FALSE}
mapHumanGS2Zebrafish <- function(x) {
  x %>% 
    as.data.frame %>% 
    set_colnames("human_entrezgene") %>%
    left_join(zebAndHumanEnsGenes, by = "human_entrezgene") %>% 
    dplyr::filter(complete.cases(.)) %>%
    dplyr::select(-human_entrezgene) %>%
    as.list %>%
    unname %>%
    .[[1]] %>%
    unique
}

h_mapped <- lapply(h_df, mapHumanGS2Zebrafish)
c1_mapped <- lapply(c1_df, mapHumanGS2Zebrafish)
c2_mapped <- lapply(c2_df, mapHumanGS2Zebrafish) 
c3_mapped <- lapply(c3_df, mapHumanGS2Zebrafish)
c4_mapped <- lapply(c4_df, mapHumanGS2Zebrafish)
c5_mapped <- lapply(c5_df, mapHumanGS2Zebrafish)
c6_mapped <- lapply(c6_df, mapHumanGS2Zebrafish)
c7_mapped <- lapply(c7_df, mapHumanGS2Zebrafish)
```

Because the mapping takes a while to run, I've saved the mapped genesets as R objects, which will now be imported:

```{r Saved-Msigdb-mapped-gene-sets}
h_mapped <- readRDS(file.path(genesetsDir, "ens_h_mapped.rds"))
c1_mapped <- readRDS(file.path(genesetsDir, "ens_c1_mapped.rds"))
c2_mapped <- readRDS(file.path(genesetsDir, "ens_c2_mapped.rds"))
c3_mapped <- readRDS(file.path(genesetsDir, "ens_c3_mapped.rds"))
c4_mapped <- readRDS(file.path(genesetsDir, "ens_c4_mapped.rds"))
c5_mapped <- readRDS(file.path(genesetsDir, "ens_c5_mapped.rds"))
c6_mapped <- readRDS(file.path(genesetsDir, "ens_c6_mapped.rds"))
c7_mapped <- readRDS(file.path(genesetsDir, "ens_c7_mapped.rds"))
```

## 3. Build indexes

We now need to build an index for each collection of gene sets. 
The index is basically the ensembl IDs of the gene sets to the corresponding row number 
in the voom object `v` previously generated in the differential gene expression analysis. 
```{r}
buildGenesetIndex <- function(x, voomObj = v){
  limma::ids2indices(x, rownames(voomObj))
}

h_idx <- buildGenesetIndex(h_mapped)
c1_idx <- buildGenesetIndex(c1_mapped)
c2_idx <- buildGenesetIndex(c2_mapped)
c3_idx <- buildGenesetIndex(c3_mapped)
c4_idx <- buildGenesetIndex(c4_mapped)
c5_idx <- buildGenesetIndex(c5_mapped)
c6_idx <- buildGenesetIndex(c6_mapped)
c7_idx <- buildGenesetIndex(c7_mapped)
```

The MSigDB indexes are ready for use in our analysis. We will also define some custom gene sets corresponding to IRE-containing genes (for details on how these were obtained, see `IREGenes/Identifying_IREGenes.Rmd`), as well as subsetting the KEGG gene sets from `c2_idx`. 

**KEGG Genesets**
```{r, eval=FALSE}
kegg_idx <- c2_idx[(names(c2_idx) %>% str_detect("KEGG"))]
```
```{r}
head(kegg_idx)
```


**IRE-containing genes**
```{r, eval=FALSE}
# Read in the IRE GRanges objects which were made from the SIRES output. 
ireUtr3 <- readRDS(here("R","IREGenes","data","ireUtr3.rds"))
ireUtr5 <- readRDS(here("R","IREGenes","data","ireUtr5.rds"))

# Extract the Ensembl ids of the 3' and 5' IRE genes and store them as a list of gene sets. 
# We will distinguish between "all" predicted IRE-containing genes and only the "high-quality" ones.
ireGenes <- list(
  ire3_all <- ireUtr3$gene_id %>% unique,
  ire5_all <- ireUtr5$gene_id %>% unique,
  ire3_hq <- ireUtr3 %>% as.data.frame %>% dplyr::filter(quality == "High") %>% use_series("gene_id") %>% unique,
  ire5_hq <- ireUtr5 %>% as.data.frame %>% dplyr::filter(quality == "High") %>% use_series("gene_id") %>% unique
) %>% set_names(c("ire3_all", "ire5_all", "ire3_hq", "ire5_hq"))

# Create index for limma functions. 
ireGenes_idx <- buildGenesetIndex(ireGenes)
```
```{r}
head(ireGenes_idx)
```


## 4. Run gene set test methods

We will be using the **fry**, **fgsea**, and **camera** methods to perform GSEA on the gene sets, and then combine the results (*p*-values) of each individual gene test using Wilkinson's method. Adjustment for multiple testing across all tests will be done using FDR and Bonferroni. The function for doing all this is in the `combinedGSEA.R` file.

```{r,eval=FALSE}
source(here("R","GSEA","combinedGSEA.R"))
```

We will run the `combinedGSEA()` function for each gene set collection, on the voom object `v` as well as the `design` and `contrasts` matrix from the DE analysis. 

```{r, eval=FALSE}
gseaResults_kegg <- combinedGSEA(v, kegg_idx, design, contrasts)
gseaResults_ire <- combinedGSEA(v, ireGenes_idx, design, contrasts)
gseaResults_h <- combinedGSEA(v, h_idx, design, contrasts)


# Export results
resultsDir <- here("R","GSEA","results")

# gseaResults_kegg %>% saveRDS(file.path(resultsDir, "gseaResults_kegg.rds"))
# gseaResults_ire %>% saveRDS(file.path(resultsDir, "gseaResults_ire.rds"))
# gseaResults_h %>% saveRDS(file.path(resultsDir, "gseaResults_h.rds"))

# gseaResults_kegg %>% write.xlsx(file.path(resultsDir, "gseaResults_kegg.xlsx"))
# gseaResults_ire %>% write.xlsx(file.path(resultsDir, "gseaResults_ire.xlsx"))
# gseaResults_h %>% write.xlsx(file.path(resultsDir, "gseaResults_h.xlsx"))

```

```{r}
gseaResults_kegg <- readRDS(file.path(resultsDir, "gseaResults_kegg.rds"))
gseaResults_ire <-  readRDS(file.path(resultsDir, "gseaResults_ire.rds"))
gseaResults_h <- readRDS(file.path(resultsDir, "gseaResults_h.rds"))
```

## 5. Proportions of up/down genes in each geneset

`fry` doesn't include this information but `mroast` does. 

```{r}
gseaResults_kegg2 <- combinedGSEA2(v, kegg_idx, design, contrasts)
gseaResults_ire2 <- combinedGSEA2(v, ireGenes_idx, design, contrasts)
gseaResults_h2 <- combinedGSEA2(v, h_idx, design, contrasts)

# gseaResults_kegg2 %>% saveRDS(file.path(resultsDir, "gseaResults_kegg2.rds"))
# gseaResults_ire2 %>% saveRDS(file.path(resultsDir, "gseaResults_ire2.rds"))
# gseaResults_h2 %>% saveRDS(file.path(resultsDir, "gseaResults_h2.rds"))
```

Rankings for mroast and fry are similar, but fry ones are smaller, likely due to the infinite rotations. 
```{r}
gseaResults_kegg2$indivTest$normoxia_6mth_mutant_vs_wt$mroast %>%
  inner_join(gseaResults_kegg$indivTest$normoxia_6mth_mutant_vs_wt$fry, 
             by = c("Geneset","NGenes"), suffix = c(".mroast", ".fry")) %>% 
  dplyr::select(Geneset,NGenes,contains("FDR.Mixed"), contains("Prop")) %>% 
  dplyr::mutate(Prop.NotSignif = 1 - PropUp - PropDown) %>%
  dplyr::arrange(FDR.Mixed.fry) %>% View
```

```{r}
gseaResults_kegg2$indivTest$normoxia_6mth_mutant_vs_wt$mroast %>%
  inner_join(gseaResults_kegg2$indivTest$normoxia_24mth_mutant_vs_wt$mroast, 
             by = c("Geneset","NGenes"), suffix = c(".6mth", ".24mth")) %>% 
  dplyr::select(Geneset,NGenes,contains("FDR.Mixed"), contains("Prop")) %>% 
  dplyr::mutate(Prop.NotSignif.6mth = 1 - PropUp.6mth - PropDown.6mth,
                Prop.NotSignif.24mth = 1 - PropUp.24mth - PropDown.24mth) %>%
  dplyr::arrange(FDR.Mixed.6mth) %>% View
```

```{r}
test<-gseaResults_kegg$combTest$normoxia_6mth_mutant_vs_wt %>% filter(fdr < 0.05)
test %>% bind_rows(gseaResults_kegg$combTest$normoxia_24mth_mutant_vs_wt %>% filter(Geneset %in% test$Geneset))


```

## 6. Butterfly plot / back-to-back stacked bar chart

This kind of chart is effective for comparisons as the y-axis is shared between two panels (facets). In this chart, we are aiming to compare the proportions of up- and down-regulated genes within each geneset for two comparisons.

Try with just the following two comparisons first:

- `normoxia_6mth_mutant_vs_wt`
- `normoxia_24mth_mutant_vs_wt`

```{r, fig.width=11}
# extract up/downregulated gene proportions from mroast
x_props <- gseaResults_h$indivTest$normoxia_6mth_mutant_vs_wt$mroast %>% 
  dplyr::select(Geneset, NGenes, PropUp, PropDown) %>% 
  dplyr::mutate(contrast="normoxia_6mth_mutant_vs_wt",
                up = NGenes * PropUp,
                down = NGenes * PropDown,
                notSig = NGenes - up - down
  ) %>%
  dplyr::select(-PropUp, -PropDown)

# join to fry results
x_fry_with_props <- gseaResults_h$indivTest$normoxia_6mth_mutant_vs_wt$fry %>%
  #dplyr::filter(FDR.Mixed < 0.05) %>%
  dplyr::select(Geneset) %>%
  left_join(x_props)

x_fry_with_props

# extract proportions from mroast
y_props <- gseaResults_h$indivTest$normoxia_24mth_mutant_vs_wt$mroast %>% 
  dplyr::select(Geneset, NGenes, PropUp, PropDown) %>% 
  dplyr::mutate(contrast="normoxia_24mth_mutant_vs_wt",
                up = NGenes * PropUp,
                down = NGenes * PropDown,
                notSig = NGenes - up - down
  ) %>%
  dplyr::select(-PropUp, -PropDown)

# join to fry results
y_fry_with_props <- gseaResults_h$indivTest$normoxia_24mth_mutant_vs_wt$fry %>%
  #dplyr::filter(FDR.Mixed < 0.05) %>%
  dplyr::select(Geneset) %>%
  left_join(y_props)

y_fry_with_props

# create the plot
# adapted from https://stackoverflow.com/questions/4559229/drawing-pyramid-plot-using-r-and-ggplot2
butterfly <- bind_rows(x_fry_with_props, y_fry_with_props) %>%
  dplyr::mutate(Geneset = gsub(x = Geneset, pattern = "_", replacement = " "),
                Geneset = gsub(x = Geneset, pattern = "HALLMARK", replacement = "")) %>%
  reshape2::melt(id.vars=c("Geneset","NGenes","contrast")) %>%
  dplyr::arrange(desc(NGenes)) %>% 
  dplyr::rename(direction = variable, ngenes = value) %>%
  dplyr::mutate(Geneset = factor(Geneset, levels = unique(Geneset)),
                direction = factor(direction, levels = c("notSig", "down", "up")),
                contrast = factor(contrast, levels = c("normoxia_6mth_mutant_vs_wt","normoxia_24mth_mutant_vs_wt"))) %>%
  ggplot(aes(x = Geneset, y = ngenes, fill = direction)) +
  geom_col() +
  ggpol::facet_share(~contrast, scales = "free") +
  scale_fill_manual(values=c("#DDDDDD","#83AF9B","#FE4365"))+
  coord_flip() +
  theme(axis.line.y = element_blank())


butterfly

# I can't figure out how to horizontally flip the left-panel with ggplot so for now
# will export this and manually adjust in powerpoint:
#export::graph2ppt(butterfly, here("R","GSEA","fig","butterfly"))

```




```{r butterflyPlotFunction}
butterflyPlot <- function(x, contrastx, contrasty, signifLevel = 1){
  # Plots a back-to-back stacked bar chart with the same y-axis using 
  # two outputs of the combinedGSEA function. 
  #
  # Args:
  #   x: the  combinedGSEA object
  #   contrastx: the name of the contrast/coef in x
  #   contrasty: the name of the contrast/coef in y
  #   signifLevel: adjusted p-value cutoff
  #
  # Returns:
  #   ggplot object
  
  # Extract proportions of up and down-regulated genes from the mroast slot. 
  x_props <- x$indivTest %>%
    .[[contrastx]] %>%
    magrittr::use_series("mroast") %>%
    dplyr::select(Geneset, NGenes, PropUp, PropDown) %>% 
    dplyr::mutate(contrast=contrastx,
                up = NGenes * PropUp,
                down = NGenes * PropDown,
                notSig = NGenes - up - down) %>%
  dplyr::select(-PropUp, -PropDown)
  
  y_props <- x$indivTest %>%
    .[[contrasty]] %>%
    magrittr::use_series("mroast") %>%
    dplyr::select(Geneset, NGenes, PropUp, PropDown) %>%
    dplyr::mutate(contrast=contrasty,
                up = NGenes * PropUp,
                down = NGenes * PropDown,
                notSig = NGenes - up - down) %>%
  dplyr::select(-PropUp, -PropDown)
  

  
  # Join mroast proportions onto the fry slot and filter for significant genesets.
  x_fry_with_props <- x$indivTest %>%
    .[[contrastx]] %>%
    magrittr::use_series("fry") %>%
    dplyr::select(Geneset) %>%
    dplyr::left_join(x_props, by = c("Geneset")) 

    y_fry_with_props <- x$indivTest %>%
    .[[contrasty]] %>%
    magrittr::use_series("fry") %>%
    dplyr::select(Geneset) %>%
    dplyr::left_join(y_props, by = c("Geneset")) 
  

  # Join the results and produce the plot
  xy <- bind_rows(x_fry_with_props, y_fry_with_props) %>%
  dplyr::mutate(Geneset = gsub(x = Geneset, pattern = "_", replacement = " "),
                Geneset = gsub(x = Geneset, pattern = "HALLMARK|KEGG", replacement = ""),
                Geneset = str_wrap(string=Geneset,width=23)) %>%
  reshape2::melt(id.vars=c("Geneset","NGenes","contrast")) %>%
  dplyr::arrange(desc(NGenes)) %>%
  dplyr::rename(direction = variable, ngenes = value) %>%
  dplyr::mutate(Geneset = factor(Geneset, levels = unique(Geneset)),
                direction = factor(direction, levels = c("notSig", "down", "up")),
                contrast = factor(contrast, levels = c(contrastx,contrasty))) %>%
  ggplot(aes(x = Geneset , y = ngenes, fill = direction)) +
  geom_col() +
  ggpol::facet_share(~contrast, scales = "free") +
  scale_fill_manual(values=c("#DDDDDD","#83AF9B","#FE4365"))+
  coord_flip() +
  theme(axis.line.y = element_blank())
  
  xy
}
```

```{r}
butterflyPlot(gseaResults_ire, contrastx = "normoxia_6mth_mutant_vs_wt", contrasty = "normoxia_24mth_mutant_vs_wt")
```

```{r fig.width=11, fig.height=20}
butterflyPlot(gseaResults_kegg, contrastx = "normoxia_6mth_mutant_vs_wt", contrasty = "normoxia_24mth_mutant_vs_wt") 
  
```

## 7. K97fs Dataset

The K97fs dataset was previously analysed with `fry` only in terms of gene set enrichment analysis. 
I decided to re-run the K97fs dataset with the current method here to see whether 

- IRE-containing genes are enriched (we expect that they will not be significantly enriched as the K97fs mutation is non AD-like).
- KEGG / Hallmark pathways are similar or not compared to those in the Q96K97 dataset for the relevant comparisons. 
- Results differ a lot between the `fry` only method and the 3-way approach using `fry`, `fgsea` and `camera`. 

I basically just re-ran the same code as above except imported the following `voom` and `design` objects in instead. 
These objects were previously saved from a `limma` DE analysis. 

```{r}
v <- readRDS(here("R","DE","data","k97_voomData.rds"))
design <- readRDS(here("R","DE","data","k97_design.rds"))
contrasts <- makeContrasts(
  levels = colnames(design), 
  K97_mutYoung_wtYoung = K97Gfs_mutant_young - K97Gfs_wt_young,
  K97_mutOld_wtOld = K97Gfs_mutant_old - K97Gfs_wt_old,
  K97_wtOld_wtYoung = K97Gfs_wt_old - K97Gfs_wt_young,
  K97_mutOld_mutYoung = K97Gfs_mutant_old - K97Gfs_mutant_young
)
```

Before running the `combinedGSEA()` function, we needed to make a change to the `voom` object involving making the rownames (ensembl gene ids) into a column so that it had the same structure as for the Q96K97 dataset and that way `combinedGSEA` runs successfully. What I really need to do sometime soon is modify the `combinedGSEA` function so that it's more robust and does checks for this sort of thing.

```{r eval=FALSE}
v$genes %<>% rownames_to_column("ensembl_gene_id")
```

After running `combinedGSEA` for the relevant indices, the results were saved as follows. 

```{r eval=FALSE}
gseaResults_kegg %>% saveRDS(here("R","GSEA","results","gseaResults_kegg_k97.rds"))
gseaResults_ire %>% saveRDS(here("R","GSEA","results","gseaResults_ire_k97.rds"))
gseaResults_h %>% saveRDS(here("R","GSEA","results","gseaResults_h_k97.rds"))
```

Quick inspection of the results indicates some differences straight up. 

- For the young mutant vs. young wild-type comparison, only the **3' IRE High Quality** (`ire3_hq`) set is enriched after Bonferroni adjustment. 

```{r}
gseaResults_ire$combTest$K97_mutYoung_wtYoung
```

- Similarly, when we look at Hallmark gene sets, the top ranked ones do not really resemble the Q96K97 set other than the MTORC1 Signalling, Hypoxia, and TNFA Signalling via NFKB. The Mitotic Spindle, G2M Checkpoint and Apoptosis gene sets being enriched in this comparison indicate there could be issues with cell division. The reason for P53 pathway being enriched is unclear at first glance, but this one wasn't enriched in the analogous comparison in the Q96K97 dataset. This suggests that while there could be shared pathways in *psen1* mutation, there also seem to be quite substantial differences. 

```{r}
gseaResults_h$combTest$K97_mutYoung_wtYoung %>% filter(bonferroni < 0.05)
```


## TODO

- [ ] Look into changing to Stouffer's method between the different enrichment methods used. Stouffer's method takes direction of change into account. We can use `qnorm` to convert *p*-values into directional z-values that can be used as input for Stouffer's method. 
- [x] Re-run `combinedGSEA` using `mroast` instead of `fry` to get the proportions of genes that are up- and down-regulated. 
- [ ] Visualisation (maybe stacked bar charts / butterfly plot) for the up/down/unchanged proportions of genes for important contrasts/comparisons. Otherwise a heatmap might work. 
- [x] Re-run `combinedGSEA` with K97fs dataset


## 5. Session Info

```{r}
sessionInfo()
```

