---
title: "Stability of IRE-containing genes"
output: 
  html_notebook: 
    toc: yes
---

```{r Packages, include=FALSE}
library(tibble)
library(reshape2)
library(magrittr)
library(dplyr)
library(AnnotationHub)
library(ggplot2)
library(ggalt)
theme_set(theme_bw())

write_fit <- function(fit, results=NULL, F.adjust="none", adjust="fdr", method="global", ...) {
    if (!is(fit, "MArrayLM")) 
        stop("fit should be an MArrayLM object")
    if (!is.null(results) && !is(results, "TestResults")) 
        stop("results should be a TestResults object")
    if (is.null(fit$t) || is.null(fit$p.value)) 
        fit <- eBayes(fit)
    method <- match.arg(method, c("separate", "global"))
    p.value <- as.matrix(fit$p.value)
    if (adjust == "none") {
        p.value.adj <- NULL
    }
    else {
        p.value.adj <- p.value
        if (method == "separate") 
            for (j in 1:ncol(p.value)) p.value.adj[, j] <- p.adjust(p.value[, 
                j], method = adjust)
        if (method == "global") 
            p.value.adj[] <- p.adjust(p.value, method = adjust)
    }
    if (F.adjust == "none" || is.null(fit$F.p.value)) 
        F.p.value.adj <- NULL
    else F.p.value.adj <- p.adjust(fit$F.p.value, method = F.adjust)
    
    tab <- list()
    tab$A <- (fit$Amean)
    tab$Coef <- (fit$coef)
    tab$t <- (fit$t)
    tab$p.value <- (p.value)
    tab$p.value.adj <- (p.value.adj)
    tab$F <- (fit$F)
    tab$F.p.value <- (fit$F.p.value)
    tab$F.p.value.adj <- F.p.value.adj
    tab$Res <- unclass(results)
    tab$Genes <- fit$genes
    tab <- data.frame(tab, check.names = FALSE)
    return(tab)
}
```


## 0. Introduction

- The purpose of this analysis is to explore whether the 
increased expression / enrichment of 3' IRE genes is due 
to increased stability of the 3' IRE transcripts. 

- We estimate changes in transcript stability 
(differential mRNA decay rate) through 
$\Delta$ logFC(mature transcript) - $\Delta$ logFC(unspliced transcript), 
as described in 
[Alkallas et al. 2017](https://www.nature.com/articles/s41467-017-00867-z).


## 1. Import Data 

- To capture information about unspliced transcript expression, we ran 
*kallisto* using a custom transcriptome index 
which includes the unspliced versions of transcripts as well 
as the mature ones. 

- We will import the *kallisto* output data as follows:

```{r import-data, eval=FALSE}
# List of directories containing kallisto results
phoenixDir <- file.path("/Volumes/biohub/20170906_Morgan_Hypoxia")
kallistoResults <- file.path(phoenixDir, "3_quantifiedData/including_unspliced")%>% 
  file.path %>%
  list.dirs 
kallistoResults <- kallistoResults[-1]

# Import the kallisto results from the list of directories. 
# Takes a while to run so we will load in a saved version. 
# importedKallisto <- edgeR::catchKallisto(kallistoResults)

importedKallisto <- readRDS(here::here("R/Velocity/data/importedKallisto.rds"))
colnames(importedKallisto$counts) <- basename(colnames(importedKallisto$counts))
```
```{r preview-kallisto}
str(importedKallisto)
```

## 2. Sample and transcript/gene metadata

### Samples

- We can match the sample names in `importedKallisto$counts` to the names used 
in the rest of the analysis so far. We will do this by importing in the 
sample metadata from the DGEList prepared in earlier parts of the analysis.

```{r import-dge, eval=FALSE}
q96_dge <- readRDS(here::here("R","DE","data","dge_g_filtered.rds"))

samples <- q96_dge$samples %>%
  mutate(path = basename(path))

# Match the order and replace sample names with more readable ones. 
counts_new <- importedKallisto$counts[, samples$path] %>% 
  set_colnames(samples$sample)

```

### Annotation

- We will obtain gene and transcript annotation from AnnotationHub:

```{r eval=FALSE}
ah <- AnnotationHub()
# ah %>%
#   subset(grepl("rerio", species)) %>%
#   subset(rdataclass == "EnsDb")
ensDb <- ah[["AH64906"]]

genes <- genes(ensDb) %>% as.data.frame()
transcripts <- transcripts(ensDb) %>% as.data.frame()
```

- Match the annotation to the ordering of rownames (transcript ID in the 
kallisto object. 

```{r}
annot <- importedKallisto$annotation %>% 
  as.data.frame %>%
  rownames_to_column("tx_id_version") %>%
  left_join(transcripts, by = "tx_id_version") %>%
  left_join(genes)
head(annot)
```

## 3. Collapsing transcript counts to gene-level counts

- *Kallisto* also calculates the overdispersions for 
transcripts, so we will take these into account when 
considering the transcript counts:

```{r}
#counts <- counts_new / importedKallisto$annotation$Overdispersion
counts <- counts_new
```

- Multiple mature transcripts may correspond to the 
same gene, so we will need to sum the counts of these 
transcripts that match the same gene. 

- Note there is no need to do the same for the unspliced versions 
as these are already at gene-level. 

```{r eval=FALSE}
mature_tx_gene_level_counts <- counts %>% 
  as.data.frame %>% 
  rownames_to_column("tx_id_version") %>%
  # Don't include the unspliced ones here as they are already at gene-level
  dplyr::filter(!grepl("_unspliced",tx_id_version)) %>%  
  left_join(annot[, c("tx_id_version", "gene_id")]) %>%
  melt %>%  
    dplyr::rename(sample = variable, count = value) %>% 
  dcast(gene_id ~ sample,
        value.var = "count",
        fun.aggregate = sum) %>%
  column_to_rownames("gene_id")

dim(mature_tx_gene_level_counts)  # 25,907 genes
```

- The final gene-level counts will be stored in `gene_level_counts`. 

```{r eval=FALSE}
unspliced_tx_gene_level_counts <- counts %>% 
  as.data.frame %>% 
  rownames_to_column("gene_id") %>%
  dplyr::filter(grepl("_unspliced",gene_id))%>%
  column_to_rownames("gene_id")

gene_level_counts <- rbind(mature_tx_gene_level_counts, 
                           unspliced_tx_gene_level_counts)

# saveRDS(gene_level_counts, here("R/Velocity/data/gene_level_counts.rds"))
```

## 4. DGEList

- To set up a DGEList for DE gene expression analysis, we need the counts, 
samples, and gene metadata:

```{r eval=FALSE}
annot_new <- data.frame(gene_id = rownames(gene_level_counts)) %>%
  left_join(genes, by = "gene_id")%>%
  dplyr::distinct(gene_id)

dge <- DGEList(
  counts = gene_level_counts, 
  genes = annot_new,
  samples = samples
) %>%
  calcNormFactors("TMM")

# saveRDS(dge, here::here("R/Velocity/data/dge.rds"))

#dge <- readRDS(here::here("R/Velocity/data/dge.rds"))
```

## 5. Filtering

- We will filter out low-expressed genes as part of the DE gene analysis. 
Here we use the threshold at least 1 cpm in four samples (the number of 
samples in each condition). 

```{r}
keepTheseGenes <- (rowSums(cpm(dge) > 1) >= 4) 

A <- dge %>% 
  cpm(log = TRUE) %>% 
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("A. Before filtering") +
  labs(x = "logCPM", y = "Density")

B <- dge %>% 
  cpm(log = TRUE) %>% 
  magrittr::extract(keepTheseGenes,) %>%
  melt %>% 
  dplyr::filter(is.finite(value)) %>% 
  ggplot(aes(x = value, colour = Var2)) +
  geom_density() + 
  guides(colour = FALSE) +
  ggtitle("B. After filtering")+
  labs(x = "logCPM", y = "Density")

grid.newpage()
vp1 <- viewport(x = 0, y = 0, width = 0.5, height = 1, just = c(0, 0))
vp2 <- viewport(x = 0.5, y = 0, width = 0.5, height = 1, just = c(0,0))
print(A, vp = vp1)
print(B, vp  = vp2)
dge <- dge[keepTheseGenes,,keep.lib.sizes = FALSE] 
```

## 6. DE Gene Analysis

- Our DE gene analysis approach will be *limma*-based. 
- The following code follows the same approach as that used previously in 
mature-transcript-only-based DE gene analysis for this dataset. 

```{r eval=FALSE}
design <- model.matrix(~0 + Group + Gender, data = dge$samples) %>%
  set_colnames(gsub(pattern = "Group", replacement = "", x = colnames(.)))

voomData <- voomWithQualityWeights(dge, design = design, plot = TRUE)

contrasts <- makeContrasts(
  levels = colnames(design), 
  
  # Effect of the mutation when hypoxic stress condition and age are constant:
  normoxia_6mth_mutant_vs_wt = q96_6_0-wt_6_0,
  normoxia_24mth_mutant_vs_wt = q96_24_0-wt_24_0,
  hypoxia_6mth_mutant_vs_wt = q96_6_1-wt_6_1,
  hypoxia_24mth_mutant_vs_wt = q96_24_1-wt_24_1,
  
  # Effect of hypoxic stress condition when mutation and age are constant:
  wt_6mth_hypoxia_vs_normoxia = wt_6_1-wt_6_0,
  mutant_6mth_hypoxia_vs_normoxia = q96_6_1-q96_6_0,
  wt_24mth_hypoxia_vs_normoxia = wt_24_1-wt_24_0,
  mutant_24mth_hypoxia_vs_normoxia = q96_24_1-q96_24_0,
  
  # Effect of aging when mutation and hypoxic stress condition are constant:
  normoxia_wtAging = wt_24_0-wt_6_0,
  hypoxia_wtAging = wt_24_1-wt_6_1,
  normoxia_mutantAging = q96_24_0-q96_6_0,
  hypoxia_mutantAging = q96_24_1-q96_6_1
)

fit <- lmFit(voomData, design) %>%
  contrasts.fit(contrasts) %>%
  eBayes(robust = TRUE)

results <- decideTests(fit, 
                       p.value = 0.05, 
                       adjust.method = "fdr", 
                       method = "global")

write_fit <- write_fit(fit, results, method = "global", adjust = "fdr")
```

- The number of DE genes for each comparison is shown below. Here we define 
a gene as being DE if the FDR-adjusted *p*-value is < 0.05. 

```{r}
summary(results)
```


## 7. Determine stability

- The measure of stability we will be using is 
$\Delta$ logFC(mature transcript) - $\Delta$ logFC(unspliced transcript), 
as demonstrated in 
[Alkallas et al. 2017](https://www.nature.com/articles/s41467-017-00867-z).

- First we need to get the logFC of mature and unspliced transcripts 
into a form we can work with. Below, we add a new column indicating whether 
the gene is spliced or unspliced. 

```{r eval=FALSE}
limma_results <- write_fit %>% 
  as.data.frame %>%
  set_rownames(rownames(voomData))%>%
  rownames_to_column("id") %>%
  mutate(unspliced = gsub(x = id, 
                          pattern = "^.*(_unspliced)$", 
                          replacement = "TRUE")) %>% 
  mutate(unspliced = gsub(x = unspliced, 
                          pattern=".*[[:digit:]]$", 
                          replacement= "FALSE")) %>% 
  mutate(unspliced = as.factor(unspliced)) %>% 
  dplyr::mutate(id = gsub(x = id, 
                          pattern = "_unspliced", 
                          replacement = "")) %>%
  inner_join(genes[, c("gene_id", "gene_name", "description")], 
             by = c("id"="gene_id")) 
```

- Extract the DE results for mature and unspliced genes separately. 

```{r eval=FALSE}
unspliced_genes <- limma_results %>% 
  dplyr::filter(unspliced == TRUE) %>%
  dplyr::select(id, starts_with("Coef")) 
colnames(unspliced_genes) %<>% gsub(x = colnames(unspliced_genes), 
                                 pattern = "Coef", 
                                 replacement = "LogFCUnspliced")

mature_genes <- limma_results %>% 
  dplyr::filter(unspliced == FALSE) %>%
  dplyr::select(id, starts_with("Coef")) 
colnames(mature_genes) %<>% gsub(x = colnames(mature_genes), 
                                 pattern = "Coef", 
                                 replacement = "LogFCMature")

allLogFC <- full_join(mature_genes, unspliced_genes, by = "id") 

# saveRDS(allLogFC, here::here("R/Velocity/data/allLogFC.rds"))
```
```{r}
head(allLogFC)
```

- Subset the DE results to be based on the mature genes only, but then 
join this to the logFC of mature and unspliced genes. This is because 
later we will be filtering based on *t* statistic / logFC of the mature 
transcripts to see whether they are stabilised or not. 

```{r}
limma_results_mature <- limma_results %>% 
  dplyr::filter(unspliced == FALSE) %>%
  left_join(allLogFC, by = "id")
```

- The `limma_results_mature` data.frame contains all comparisons in one so 
we will separate this out into a list of data.frames for each comparison 
(`coef`). 

- We will also create a new column `diff` which is the difference between 
the logFCs of mature and unspliced transcripts. 

- We define a gene as being "stabilised" in a particular comparison if 
`diff` > 0. (**Need to check with Steve**). 

- In addition, we will filter for 3' IRE genes to see if increased expression 
(defined as t >= 2) corresponds to stabilisation (`diff` > 0). 

```{r eval=FALSE}
# List of list of IRE-containing genes with Ensembl IDs
# All predicted 3' IRE genes, All predicted 5' IRE genes
# High-quality 3' IRE genes, and High-quality 5' IRE genes
ireGenes <- readRDS(here::here("R/IREGenes/data/zebrafishIreGenes.rds"))

# Names of comparisons from contrasts
coefs <- c("normoxia_6mth_mutant_vs_wt", "normoxia_24mth_mutant_vs_wt", 
           "hypoxia_6mth_mutant_vs_wt", "hypoxia_24mth_mutant_vs_wt",
           "wt_6mth_hypoxia_vs_normoxia", "mutant_6mth_hypoxia_vs_normoxia",
           "wt_24mth_hypoxia_vs_normoxia", "mutant_24mth_hypoxia_vs_normoxia",
           "normoxia_wtAging", "hypoxia_wtAging",
           "normoxia_mutantAging", "hypoxia_mutantAging")

# Store results for each comparison in its own data.frame. 
indivResults <- coefs %>% 
  lapply(function(coef){
  limma_results_mature %>% 
      dplyr::select(id, 
                    gene_name, 
                    description, 
                    contains(coef))%>%
    set_colnames(c("id", 
                   "gene_name", 
                   "description", 
                   "coef", "t", "p", "fdr", "res",  
                   "logfc_mature", "logfc_unspliced")) %>%
    dplyr::mutate(diff = logfc_mature - logfc_unspliced,
                  stabilised = diff > 0) %>%
    dplyr::filter(coef > 0 & t >= 2 & id %in% ireGenes$ire3_all) 
})
names(indivResults) <- coefs

indivResults %>% lapply()
```
```{r}
View(indivResults$normoxia_6mth_mutant_vs_wt)
```


## 8. Significance of results

- Here I used the Mann-Whitney U-test to test for significant 
differences in `mature_logfc` and `unspliced_logfc`, as the distributions of 
`mature_logfc` and `unspliced_logfc` do not look normal. 

- **Null hypothesis (no stabilisation)**: `mature_logfc` and `unspliced_logfc` 
are the same. (True location shift = 0)

- **Alternative hypothesis (mature transcripts are stabilised)**: `mature_logfc` 
is greater than `unspliced_logfc`. (True location shift > 0)

- I'm not sure whether this is entirely correct but I've specified 
`paired = TRUE` for `mature_logfc` and `unspliced_logfc` from the same gene. I'm 
not sure whether these should be said to be independent or dependent but I was 
thinking they might be dependent? 

- All p-values FDR-adjusted and we define significance level to be 0.05. 

The results below indicate whether 3' IRE-containing genes that are 
increased in expression are significantly more stable in each comparison.

```{r}
tests <- indivResults %>% lapply(function(x) {
  wilcox.test(x$logfc_mature, 
              x$logfc_unspliced, 
              paired=TRUE, 
              alternative = "greater")
})

tests %>% lapply(function(x){
  x$p.value
}) %>%
  as.data.frame %>% 
  t %>%
  as.data.frame %>%
  set_colnames("p") %>%
  rownames_to_column("comparison") %>%
  mutate(fdr = p.adjust(p, method = "fdr"),
         sig = fdr < 0.05)%>%View

```



## 8. Visualisation

- To visualise these results, I'm trying out dumbbell plots to show 
the difference between the mature logFC and unspliced logFC visually. 

- Red = Unspliced logFC; Blue = Mature logFC. 

**normoxia_6mth_mutant_vs_wt**:

```{r}
indivResults$normoxia_6mth_mutant_vs_wt %>%
  arrange(diff) %>%
  dplyr::filter(!is.na(diff)) %>%
  dplyr::mutate(gene_name = factor(gene_name, 
                                   levels = as.character(gene_name))) %>%
  ggplot(aes(x = logfc_mature, xend = logfc_unspliced, y = gene_name, 
             group = gene_name)) +
  geom_dumbbell(colour_x="cornflowerblue",
                  colour_xend = "orangered",
                  colour = "grey50",
                dot_guide = TRUE,
                  size_x = 1.5, size_xend=1.5) + 
    scale_x_continuous() +
    labs(x=NULL, 
         y=NULL) +
    theme(plot.title = element_text(hjust=0.5, face="bold"),
          aspect.ratio = 1,
          panel.grid.minor=element_blank(),
          panel.grid.major.y=element_blank(),
          panel.grid.major.x=element_line(),
          axis.ticks=element_blank(),
          #legend.position="bottom",
          panel.border=element_blank(),
          axis.text.y = element_text(lineheight = 10,size = 6)) +
  geom_hline(yintercept=9.5, colour = "gray50", size=1)
  
```


```{r}
indivResults$normoxia_24mth_mutant_vs_wt %>%
  arrange(diff) %>%
  dplyr::filter(!is.na(diff)) %>%
  dplyr::mutate(gene_name = factor(gene_name, 
                                   levels = as.character(gene_name))) %>%
  ggplot(aes(x = logfc_mature, xend = logfc_unspliced, y = gene_name, 
             group = gene_name)) +
  geom_dumbbell(colour_x="cornflowerblue",
                  colour_xend = "orangered",
                  colour = "grey50",
                dot_guide = TRUE,
                  size_x = 1.5, size_xend=1.5) + 
    scale_x_continuous() +
    labs(x=NULL, 
         y=NULL) +
    theme(plot.title = element_text(hjust=0.5, face="bold"),
          aspect.ratio = 1,
          panel.grid.minor=element_blank(),
          panel.grid.major.y=element_blank(),
          panel.grid.major.x=element_line(),
          axis.ticks=element_blank(),
          legend.position="top",
          panel.border=element_blank(),
          axis.text.y = element_text(lineheight = 10,size = 6)) +
  geom_hline(yintercept=7.5, colour = "gray50", size=1)
```

```{r fig.height=12}
indivResults$wt_6mth_hypoxia_vs_normoxia %>%
  arrange(diff) %>%
  dplyr::filter(!is.na(diff)) %>% 
  dplyr::mutate(gene_name = factor(gene_name, 
                                   levels = as.character(gene_name))) %>%
  ggplot(aes(x = logfc_mature, xend = logfc_unspliced, y = gene_name, 
             group = gene_name)) +
  geom_dumbbell(colour_x="cornflowerblue",
                  colour_xend = "orangered",
                  colour = "grey50",
                dot_guide = TRUE,
                  size_x = 1.5, size_xend=1.5) + 
    scale_x_continuous() +
    labs(x=NULL, 
         y=NULL) +
    theme(plot.title = element_text(hjust=0.5, face="bold"),
          aspect.ratio = 1,
          panel.grid.minor=element_blank(),
          panel.grid.major.y=element_blank(),
          panel.grid.major.x=element_line(),
          axis.ticks=element_blank(),
          legend.position="top",
          panel.border=element_blank(),
          axis.text.y = element_text(lineheight = 10,size = 6)) 
  # geom_hline(yintercept=31.5, colour = "gray50", size=1)
  # geom_hline(yintercept=31.5, colour = "gray50", size=1)
```


## 9. T-tests to determine significance of stability

- We define the null and alternate hypotheses as follows:

    - **Null**: No stabilisation => logFC(spliced) <= logFC(unspliced)

    - **Alternate**: Stabilisation => logFC(spliced) > logFC(unspliced)

- Because the two distributions of logFC(spliced) and logFC(unspliced) 
could have unequal variance, we use the Welch's t-test (unequal variances) 
t-test. 

- We use the `s2.prior` values from the limma fit object for the variances 
`s1` and `s2` for the spliced (mature) and unspliced genes. 

- After calculating the *t*-statistic and degrees of freedom for each gene, we 
can use these statistics with the *t*-distribution to test the null hypothesis.



```{r}
# Extract s2.post from the limma fit object
sDf <- data.frame(gene_id = rownames(fit), s2.post = fit$s2.post)

# Make a data.frame all_s where each row is the gene and 
# there are columns s1 and s2 for the variances of the 
# mature and unspliced version of each gene. 
sDf <- data.frame(id = rownames(fit), 
           s2.post = fit$s2.post) %>% 
  mutate(unspliced = gsub(x = id, 
                          pattern = "^.*(_unspliced)$", 
                          replacement = "TRUE")) %>% 
    mutate(unspliced = gsub(x = unspliced, 
                            pattern=".*[[:digit:]]$", 
                            replacement= "FALSE")) %>% 
    mutate(unspliced = as.factor(unspliced)) %>% 
    dplyr::mutate(id = gsub(x = id, 
                            pattern = "_unspliced", 
                            replacement = "")) %>%
    inner_join(genes[, c("gene_id", "gene_name", "description")], 
               by = c("id"="gene_id")) 
unspliced_s <- sDf %>% 
  dplyr::filter(unspliced == TRUE)
mature_s <- sDf %>% 
  dplyr::filter(unspliced == FALSE)
all_s <- full_join(mature_s, unspliced_s, 
                   by = c("id","gene_name","description")) %>%
  dplyr::select(-unspliced.x, -unspliced.y, -description) %>%
  dplyr::rename(s_mature = s2.post.x, s_unspliced = s2.post.y)


tResults <- indivResults %>%
  lapply(function(x){
    x %>% 
      left_join(all_s, by = c("id", "gene_name")) %>%
      mutate(
        tstat = (logfc_mature - logfc_unspliced) / sqrt((s_mature^2 + s_unspliced^2)/32),
        df = (32-1)*(s_mature^2+s_unspliced^2)^2 / (s_mature^4 + s_unspliced^4)
        # df = fit$df.total[1]
      )  %>%
      mutate(
        pt = pt(tstat, df, lower.tail = FALSE),
        fdr = p.adjust(pt, "fdr")
      ) %>%
      dplyr::select(
        id, gene_name,diff, tstat, df, pt, fdr,
        starts_with("logfc"), starts_with("s_")
      ) %>%
      arrange(pt)
  }) %>%
  lapply(as_tibble)



# Calculate t statistic for sample sample size (n=32) & unequal variance
# and df for each gene. 
# The pt() function allows us to calculate the distribution function 
# based on the t statistic and degrees of freedom. 
calc <- all_s %>% mutate(t = (logfc_mature - logfc_unspliced) / 
                              sqrt((s1^2 + s2^2)/32),
                 df = (32-1)*(s1^2+s2^2)^2 / (s1^4 + s2^4) )  %>%
  rowwise() %>%
  mutate(pt = pt(t, df))
```

```{r}

# pt(calc$t, calc$df)
# agingTest <- indivResults$normoxia_wtAging%>% 
#   dplyr::select(id, gene_name, contains("logfc")) %>%
#   left_join(all_s, by = c("id", "gene_name")) %>%
#   dplyr::rename(s_mature = s1, s_unspliced = s2) %>%
#   mutate(t = (logfc_mature - logfc_unspliced) / 
#                               sqrt((s_mature^2 + s_unspliced^2)/32),
#                  df = (32-1)*(s_mature^2+s_unspliced^2)^2 / 
#                       (s_mature^4 + s_unspliced^4) )  %>%
#   rowwise() %>%
#   mutate(pt = pt(t, df))

```

