---
title: "Neural cell type markers"
output: html_notebook
---
```{r}
library(here)
mouse_dge <- readRDS(here("R/Mouse/data/dge.rds"))
q96_dge <- readRDS(here("R/DE/data/dge_g_filtered.rds"))
human_dge <- readRDS(here("R/GSEA/data/human/mayo_DGE.rds"))

# options
theme_set(theme_bw())
```


## 0. Introduction

- Different proportions of cell types can result in DE gene changes even 
if the genes themselves aren't changing in expression. 

- This is just a quick test to see whether cell type proportions are 
relatively the same across different conditions. 

- We will do this for all datasets analysed in the paper. 

## 1. Cell type markers

- The C2 collection contains some gene sets which were found to be markers of 
neural cell types based on *in vivo* studies. 

- The relevant gene sets are:

    - `LEIN_ASTROCYTE_MARKERS`
    - `LEIN_NEURON_MARKERS`
    - `LEIN_OLIGODENDROCYTE_MARKERS`
  
- From previous GSEA analyses, we have downloaded the GSEA gene set collections 
from MSigDB and converted all gene IDs to Ensembl gene IDs. We now 
import in the C2 collection for each species:

```{r}
human_c2_mapped <- readRDS(here("R", "GSEA", "data", "human", "ens_c2_mapped.rds"))
mouse_c2_mapped <- readRDS(here("R/GSEA/data/human/ens_c2_mapped_m.rds"))
zeb_c2_mapped <- readRDS(here("R/GSEA/data/ens_c2_mapped.rds"))
```

- Couldn't find any MSigDB gene set corresponding to microglia, but this 
[paper](https://www.frontiersin.org/articles/10.3389/fimmu.2019.01170/full#SM1) 
has a list of them:

```{r}
human_microglia <- c("CD33", "CD84", "FYB1", "FYB2", "FCER1G", 
                     "GPR34", "ACY3", "ALOX5AP", "CSF2RA", "C1QB", 
                     "C3", "IGSF6", "LY86", "LAPTM5", "P2RY13", "SPN",
                     "SUCNR1", "TBXAS1", "TLR7", "TREM2")

library(biomaRt)
ensembl_human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
ensembl_mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
ensembl_zeb <- useMart("ensembl", dataset = "drerio_gene_ensembl")

human_to_mouse_microglia <- getBM(
  attributes = c("mmusculus_homolog_associated_gene_name", "mmusculus_homolog_chromosome"),
  filters = "external_gene_name",
  values = human_microglia, 
  mart = ensembl_human
) %>%
  inner_join(mouse_dge$genes[, c("symbol", "ensembl_gene_id")], 
            by = c("mmusculus_homolog_associated_gene_name"="symbol")) 
mouse_microglia = human_to_mouse_microglia$ensembl_gene_id

human_to_zeb_microglia <- getBM(
  attributes = c("drerio_homolog_associated_gene_name", "drerio_homolog_chromosome"),
  filters = "external_gene_name",
  values = human_microglia, 
  mart = ensembl_human
) %>%
  dplyr::filter(drerio_homolog_associated_gene_name != "") %>%
  inner_join(q96_dge$genes[, c("gene_name", "ensembl_gene_id")], 
            by = c("drerio_homolog_associated_gene_name"="gene_name")) 
zeb_microglia = human_to_zeb_microglia$ensembl_gene_id

human_microglia <- human_dge$genes %>% 
  rownames_to_column("ensembl_gene_id") %>%
  dplyr::filter(external_gene_name %in% human_microglia) %>%
  .$ensembl_gene_id

```


- For each species, we will extract the relevant gene sets:

```{r}
human_neuralMarkers <- list(
  astrocyte = human_c2_mapped$LEIN_ASTROCYTE_MARKERS,
  neuron = human_c2_mapped$LEIN_NEURON_MARKERS,
  oligodendrocyte = human_c2_mapped$LEIN_OLIGODENDROCYTE_MARKERS,
  microglia = human_microglia
)

mouse_neuralMarkers <- list(
  astrocyte = mouse_c2_mapped$LEIN_ASTROCYTE_MARKERS,
  neuron = mouse_c2_mapped$LEIN_NEURON_MARKERS,
  oligodendrocyte = mouse_c2_mapped$LEIN_OLIGODENDROCYTE_MARKERS,
  microglia = mouse_microglia
)

zeb_neuralMarkers <- list(
  astrocyte = zeb_c2_mapped$LEIN_ASTROCYTE_MARKERS,
  neuron = zeb_c2_mapped$LEIN_NEURON_MARKERS,
  oligodendrocyte = zeb_c2_mapped$LEIN_OLIGODENDROCYTE_MARKERS,
  microglia = zeb_microglia
)
```

## 2. Import Datasets

- The human datasets we analysed included:

    - Caco-2 (intestinal) cell line culture - microarray (doesn't need to be analysed here as it's a cell line)
    - Mayo Clinic study - RNA-seq 

- Zebrafish datasets:

    - fAD-like mutant zebrafish (Q96_K97del/+) - RNAseq
    - non fAD-like mutant zebrafish (K97fs/+) - RNA-seq

- Mouse dataset:

    - 5xFAD mice - RNA-seq
    
- We will import the expression values / count matrix for these datasets 
below. 

```{r}
# Human
human_dge <- readRDS(here("R/GSEA/data/human/mayo_DGE.rds"))
#caco2_exprs <- readRDS(here("R/IronDepleted/data/cellLine_3573/genesSamplesCounts.rds"))
# Zebrafish
q96_dge <- readRDS(here("R/DE/data/dge_g_filtered.rds"))
k97_dge <- readRDS(here("R/DE/data/k97_dge.rds"))
# Mouse
mouse_dge <- readRDS(here("R/Mouse/data/dge.rds"))
```

- The next step is to subset the expression for the gene sets. We will then 
use this to produce visualisations of the expression values.

### Human

- In the case of the Mayo Clinic RNA_seq dataset, we have many ages. 
Age greatly influences gene expression,so rather than a heatmap, we will 
show the impact of age as well as the diagnosis through trendlines on a 
scatterplot. 

- The plots below show expression in the **temporal cortex**. This region is 
known to be more affected than the cerebellum in neurodegeneration. 

- **Summary**: Microglia and astrocytes show strong differences but direction 
isn't always consistent, so we can't conclude if the number of cells is 
changing greatly or if it's just the expression changing. For neuron and 
oligodendrocyte marker genes, differences are more subtle and may not be. 


```{r }
# for facet graphs, set fig.width=20
human_exp <- lapply(human_neuralMarkers, function(x){
  human_dge %>% 
    cpm(log=TRUE) %>%
    as.data.frame%>%
    rownames_to_column("ensembl") %>% 
    dplyr::filter(ensembl %in% x) %>%
    left_join(human_dge$genes %>% as.data.frame %>%rownames_to_column("ensembl") %>% dplyr::select(external_gene_name, ensembl)) %>%
    dplyr::distinct(external_gene_name, .keep_all = TRUE) %>%
    dplyr::select(-ensembl) %>%
    as.data.frame %>%
    #column_to_rownames("external_gene_name") %>%
    melt %>%
    left_join(human_dge$samples, by = c("variable"="SampleID"))
})

# Astrocyte marker genes in temporal cortex
plot_human_astrocyte <- human_exp$astrocyte %>% 
  dplyr::filter(Tissue=="TemporalCortex",
                Diagnosis %in% c("Control", "AD"))%>%
  ggplot(aes(x = AgeAtDeath, y = value, colour = Diagnosis)) +
  geom_point(size = 0.5) +
  geom_smooth(aes(fill = Diagnosis), alpha = 0.2) +
  scale_colour_manual(values = c("gray50", "red", "DodgerBlue", "MediumSeaGreen")) +
  scale_fill_manual(values = c("gray50","red","DodgerBlue", "MediumSeaGreen")) +
  facet_wrap(~external_gene_name, scales = "free_y") +
  labs(x = "Age", y = "Gene expression (log2 cpm)")
plot_human_astrocyte
#saveRDS(plot_human_astrocyte, here("R/CellTypeProportions/fig/plot_human_astrocyte_TC.rds"))
plot_human_astrocyte <- human_exp$astrocyte %>% 
  dplyr::filter(Tissue=="TemporalCortex",
                Diagnosis %in% c("Control", "AD"))%>%
  ggplot(aes(x = AgeAtDeath, y = value, colour = Diagnosis)) +
  #geom_point(size = 0.5) +
  geom_point(size = 0.2, alpha=0.05) +
  geom_smooth(aes(fill = Diagnosis), alpha = 0.3) +
  scale_colour_manual(values = c("gray50", "red", "DodgerBlue", "MediumSeaGreen")) +
  scale_fill_manual(values = c("gray50","red","DodgerBlue", "MediumSeaGreen")) +
  #facet_wrap(~external_gene_name, scales = "free_y") +
  labs(x = "Age", y = "Gene expression (log2 cpm)")
plot_human_astrocyte

# Microglial marker genes in temporal cortex
plot_human_microglia <- human_exp$microglia %>% 
  dplyr::filter(Tissue=="TemporalCortex",
                Diagnosis %in% c("Control","AD"))%>%
  ggplot(aes(x = AgeAtDeath, y = value, colour = Diagnosis)) +
  #geom_point(size = 0.5) +
  geom_point(size = 0.2, alpha=0.05) +
  geom_smooth(aes(fill = Diagnosis), alpha = 0.3) +
  scale_colour_manual(values = c("gray50", "red", "DodgerBlue", "MediumSeaGreen")) +
  scale_fill_manual(values = c("gray50","red","DodgerBlue", "MediumSeaGreen")) +
  #facet_wrap(~external_gene_name, scales = "free_y") +
  labs(x = "Age", y = "Gene expression (log2 cpm)")
plot_human_microglia
#saveRDS(plot_human_microglia, here("R/CellTypeProportions/fig/plot_human_microglia_TC.rds"))

# Neuron marker genes in temporal cortex
plot_human_neuron <- human_exp$neuron %>% 
    dplyr::filter(Tissue=="TemporalCortex",
                Diagnosis %in% c("Control","AD"))%>%
  ggplot(aes(x = AgeAtDeath, y = value, colour = Diagnosis)) +
  #geom_point(size = 0.5) +
  geom_point(size = 0.2, alpha=0.05) +
  geom_smooth(aes(fill = Diagnosis), alpha = 0.3) +
  scale_colour_manual(values = c("gray50", "red", "DodgerBlue", "MediumSeaGreen")) +
  scale_fill_manual(values = c("gray50","red","DodgerBlue", "MediumSeaGreen")) +
  #facet_wrap(~external_gene_name, scales = "free_y") +
  labs(x = "Age", y = "Gene expression (log2 cpm)")
#saveRDS(plot_human_neuron, here("R/CellTypeProportions/fig/plot_human_neuron_TC.rds"))
plot_human_neuron

# Oligodendrocyte marker genes in temporal cortex
plot_human_oligodendrocyte <- human_exp$oligodendrocyte %>% 
      dplyr::filter(Tissue=="TemporalCortex",
                Diagnosis %in% c("Control","AD"))%>%
  ggplot(aes(x = AgeAtDeath, y = value, colour = Diagnosis)) +
  geom_point(size = 0.2, alpha=0.05) +
  scale_shape_manual(values = c(1)) +
  geom_smooth(aes(fill = Diagnosis), alpha = 0.3) +
  scale_colour_manual(values = c("gray50", "red", "DodgerBlue", "MediumSeaGreen")) +
  scale_fill_manual(values = c("gray50","red","DodgerBlue", "MediumSeaGreen")) +
  #facet_wrap(~external_gene_name, scales = "free_y") +
  labs(x = "Age", y = "Gene expression (log2 cpm)")
#saveRDS(plot_human_oligodendrocyte, here("R/CellTypeProportions/fig/plot_human_oligodendrocyte_TC.rds"))
plot_human_oligodendrocyte
```

### Mouse

- The samples used for RNAseq are from the cortex region of the mice. 
- There are 3 age groups: 3 months, 6 months, and 12 months. 
- I think that a similar thing to the humans would work alright. 

```{r message=FALSE, warning=FALSE}
mouse_long <- mouse_dge %>% cpm(log=TRUE) %>%
  as.data.frame %>%
  rownames_to_column("ensembl_gene_id") %>%
  left_join(mouse_dge$genes[, c("gene_name", "ensembl_gene_id")]) %>%
  melt%>%
  dplyr::rename(id = variable) %>%
  left_join(mouse_dge$samples, by = "id") 

mouse_long_exp <- lapply(mouse_neuralMarkers, function(x){
  mouse_long %>% dplyr::filter(ensembl_gene_id %in% x)
})



plot_mouse_astrocyte <- mouse_long_exp$astrocyte %>%
  ggplot(aes(x = as.factor(age), y = value,fill=genotype)) +
  geom_boxplot() +
  #scale_colour_manual(values = c("red","gray50")) +
  scale_fill_manual(values = c("red", "gray50")) +
  #facet_wrap(~gene_name, scales = "free_y") +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
#saveRDS(plot_mouse_astrocyte, here("R/CellTypeProportions/fig/plot_mouse_astrocyte.rds"))
plot_mouse_astrocyte

plot_mouse_neuron <- mouse_long_exp$neuron %>%
  ggplot(aes(x = as.factor(age), y = value,fill=genotype)) +
  geom_boxplot() +
  #scale_colour_manual(values = c("red","gray50")) +
  scale_fill_manual(values = c("red", "gray50")) +
  #facet_wrap(~gene_name, scales = "free_y") +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
#saveRDS(plot_mouse_astrocyte, here("R/CellTypeProportions/fig/plot_mouse_astrocyte.rds"))
plot_mouse_neuron
# 
# plot_mouse_neuron <- mouse_long_exp$neuron %>%
#   ggplot(aes(x = age, y = value, colour = genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = genotype), alpha=0.2, size=0.5) +
#   scale_colour_manual(values = c("red","gray50")) +
#   scale_fill_manual(values = c("red", "gray50")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_mouse_neuron, here("R/CellTypeProportions/fig/plot_mouse_neuron.rds"))
# plot_mouse_neuron

# plot_mouse_oligodendrocyte <- mouse_long_exp$oligodendrocyte %>%
#   ggplot(aes(x = age, y = value, colour = genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = genotype), alpha=0.2, size=0.5) +
#   scale_colour_manual(values = c("red","gray50")) +
#   scale_fill_manual(values = c("red", "gray50")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_mouse_oligodendrocyte, here("R/CellTypeProportions/fig/plot_mouse_oligodendrocyte.rds"))
# plot_mouse_oligodendrocyte
plot_mouse_oligodendrocyte <- mouse_long_exp$oligodendrocyte %>%
  ggplot(aes(x = as.factor(age), y = value,fill=genotype)) +
  geom_boxplot() +
  #scale_colour_manual(values = c("red","gray50")) +
  scale_fill_manual(values = c("red", "gray50")) +
  #facet_wrap(~gene_name, scales = "free_y") +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
#saveRDS(plot_mouse_astrocyte, here("R/CellTypeProportions/fig/plot_mouse_astrocyte.rds"))
plot_mouse_oligodendrocyte

# plot_mouse_microglia <- mouse_long_exp$microglia %>%
#   ggplot(aes(x = age, y = value, colour = genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = genotype), alpha=0.2, size=0.5) +
#   scale_colour_manual(values = c("red","gray50")) +
#   scale_fill_manual(values = c("red", "gray50")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_mouse_microglia, here("R/CellTypeProportions/fig/plot_mouse_microglia.rds"))
# plot_mouse_microglia


plot_mouse_microglia <- mouse_long_exp$microglia %>%
  ggplot(aes(x = as.factor(age), y = value,fill=genotype)) +
  geom_boxplot() +
  #scale_colour_manual(values = c("red","gray50")) +
  scale_fill_manual(values = c("red", "gray50")) +
  #facet_wrap(~gene_name, scales = "free_y") +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
#saveRDS(plot_mouse_astrocyte, here("R/CellTypeProportions/fig/plot_mouse_astrocyte.rds"))
plot_mouse_microglia
```

### Zebrafish

- The experimental design is more complex here. We only have two age groups, 
but also hypoxia as well. Might make a new factor `age_hypoxia`. 

```{r message=FALSE, warning=FALSE}
q96_long <- q96_dge$counts %>% cpm(log = TRUE) %>%
  as.data.frame %>%
  rownames_to_column("ensembl_gene_id") %>%
  left_join(q96_dge$genes[, c("ensembl_gene_id", "gene_name")]) %>%
  melt %>%
  dplyr::rename(sample=variable) %>%
  left_join(q96_dge$samples, by = "sample") %>%
  mutate(age_hypoxia = paste0("months_", Age, "_", Hypoxia) %>% 
           factor(levels = c("months_6_0", "months_6_1", "months_24_0", "months_24_1"))) %>%
  mutate(age_hypoxia_num = case_when(
    age_hypoxia=="months_6_0" ~ 1,
    age_hypoxia=="months_6_1" ~ 2,
    age_hypoxia=="months_24_0" ~ 3,
    age_hypoxia=="months_24_1" ~4
  )) %>%
  mutate(age_hypoxia_name = case_when(
    age_hypoxia=="months_6_0" ~ "6 months, normoxia",
    age_hypoxia=="months_6_1" ~ "6 months, hypoxia",
    age_hypoxia=="months_24_0" ~ "24 months, normoxia",
    age_hypoxia=="months_24_1" ~ "24 months, hypoxia"
  ))

q96_long_exp <- lapply(zeb_neuralMarkers, function(x){
  q96_long %>% dplyr::filter(ensembl_gene_id %in% x)
})

# plot_zeb_astrocytes <- q96_long_exp$astrocyte %>%
#   ggplot(aes(x = age_hypoxia_num, y = value, colour= Genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = Genotype), alpha=0.2, size = 0.5) +
#   scale_colour_manual(values = c("gray50","red")) +
#   scale_fill_manual(values = c("gray50", "red")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_zeb_astrocytes, here("R/CellTypeProportions/fig/plot_zeb_astrocytes.rds"))
# plot_zeb_astrocytes
plot_zeb_astrocyte <- q96_long_exp$astrocyte %>%
  ggplot(aes(x = age_hypoxia_name, y = value, fill = Genotype)) +
  geom_boxplot() +
  scale_fill_manual(values = c("gray50", "red")) +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
plot_zeb_astrocyte

# plot_zeb_neuron <- q96_long_exp$neuron %>%
#   ggplot(aes(x = age_hypoxia_num, y = value, colour= Genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = Genotype), alpha=0.2, size = 0.5) +
#   scale_colour_manual(values = c("gray50","red")) +
#   scale_fill_manual(values = c("gray50", "red")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_zeb_neuron, here("R/CellTypeProportions/fig/plot_zeb_neuron.rds"))
# plot_zeb_neuron
plot_zeb_neuron <- q96_long_exp$neuron %>%
  ggplot(aes(x = age_hypoxia_name, y = value, fill = Genotype)) +
  geom_boxplot() +
  scale_fill_manual(values = c("gray50", "red")) +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
plot_zeb_neuron

# plot_zeb_oligodendrocyte <- q96_long_exp$oligodendrocyte %>%
#   ggplot(aes(x = age_hypoxia_num, y = value, colour= Genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = Genotype), alpha=0.2, size = 0.5) +
#   scale_colour_manual(values = c("gray50","red")) +
#   scale_fill_manual(values = c("gray50", "red")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_zeb_oligodendrocyte, here("R/CellTypeProportions/fig/plot_zeb_oligodendrocyte.rds"))
# plot_zeb_oligodendrocyte

plot_zeb_oligodendrocyte <- q96_long_exp$oligodendrocyte %>%
  ggplot(aes(x = age_hypoxia_name, y = value, fill = Genotype)) +
  geom_boxplot() +
  scale_fill_manual(values = c("gray50", "red")) +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
plot_zeb_oligodendrocyte

# plot_zeb_microglia <- q96_long_exp$microglia %>%
#   ggplot(aes(x = age_hypoxia_num, y = value, colour= Genotype)) +
#   geom_point(size = 0.5) +
#   geom_smooth(aes(fill = Genotype), alpha=0.2, size = 0.5) +
#   scale_colour_manual(values = c("gray50","red")) +
#   scale_fill_manual(values = c("gray50", "red")) +
#   facet_wrap(~gene_name, scales = "free_y") +
#   labs(x = "Age", y = "Gene expression (log2 cpm)") 
# saveRDS(plot_zeb_microglia, here("R/CellTypeProportions/fig/plot_zeb_microglia.rds"))
# plot_zeb_microglia

plot_zeb_microglia <- q96_long_exp$microglia %>%
  ggplot(aes(x = age_hypoxia_name, y = value, fill = Genotype)) +
  geom_boxplot() +
  scale_fill_manual(values = c("gray50", "red")) +
  labs(x = "Age in months", y = "Gene expression (log2 cpm)") 
plot_zeb_microglia

```


```{r}
# Neural marker genes identified in mice originally. 
c2_astrocyteMarkers <- c2_mapped$LEIN_ASTROCYTE_MARKERS %>%
  as.data.frame %>%
  set_colnames("entrezgene") %>%
  left_join(zebrafishEntrezgene2Ensembl) %>%#60 genes
  filter(complete.cases(.)) %>%#49 genes
  unique %>%
  dplyr::select(ensembl_gene_id, external_gene_name)

c2_neuronMarkers <- c2_mapped$LEIN_NEURON_MARKERS %>%
  as.data.frame %>%
  set_colnames("entrezgene") %>%
  left_join(zebrafishEntrezgene2Ensembl)%>% #108 genes
  filter(complete.cases(.)) %>% #100 genes
  unique  %>%
  distinct(external_gene_name, .keep_all=TRUE)%>%
  dplyr::select(ensembl_gene_id, external_gene_name)

c2_oligodendrocyteMarkers <- c2_mapped$LEIN_OLIGODENDROCYTE_MARKERS %>%
  as.data.frame %>%
  set_colnames("entrezgene") %>%
  left_join(zebrafishEntrezgene2Ensembl) %>% #140 genes
  filter(complete.cases(.)) %>%#96 genes
  unique %>%
   distinct(external_gene_name, .keep_all=TRUE)%>%
  dplyr::select(ensembl_gene_id, external_gene_name)

incipientAD_down <- c2_mapped$BLALOCK_ALZHEIMERS_DISEASE_INCIPIENT_DN  %>%
  as.data.frame %>%
  set_colnames("entrezgene") %>%
  left_join(zebrafishEntrezgene2Ensembl) %>% #140 genes
  filter(complete.cases(.)) %>%#96 genes
  unique %>%
   distinct(external_gene_name, .keep_all=TRUE)%>%
  dplyr::select(ensembl_gene_id, external_gene_name)

incipientAD_up <- c2_mapped$BLALOCK_ALZHEIMERS_DISEASE_INCIPIENT_UP  %>%
  as.data.frame %>%
  set_colnames("entrezgene") %>%
  left_join(zebrafishEntrezgene2Ensembl) %>% #140 genes
  filter(complete.cases(.)) %>%#96 genes
  unique %>%
   distinct(external_gene_name, .keep_all=TRUE) %>%
  dplyr::select(ensembl_gene_id, external_gene_name)
```

